---
title: 'Characterize genes detected after including introns'
author: "Ramon Massoni-Badosa"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction

Here, we will perform the analysis needed to answer this concern from reviewer #3:

*"Were the scRNA-seq reads and snRNA-seq (from multiome) aligned to the same reference (e.g. transcriptome with/without introns)? This might affect the number of common features between scRNA-seq and snRNA-seq that the authors report as being very few."*

We aim to answer this question using data from 4 donors for which we had both scRNA-seq and Multiome data (BCLL-8-T, BCLL-9-T, BCLL-14-T, BCLL-15-T). We selected one GEM well per donor and technology (see metadata below), and we mapped the fastq files to obtain gene expression matrices using cellranger and cellranger-arc with and without introns. In particular, we used [cellranger 7.1](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/release-notes) for scRNA-seq data with the flag "--include-introns" set to true or false and the reference refdata-gex-GRCh38-2020-A. For multiome data, we used [cellranger-arc 2.0.2](https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/what-is-cell-ranger-arc) with or without the flag "--gex-exclude-introns" and the reference refdata-cellranger-arc-GRCh38-2020-A-2.0.0.

In this notebook, we will load all expression matrices and subset them to the cells that were included in the final atlas. We will also apply the same gene filter cutoff as applied in the atlas. Then, we will assess the following:

- Changes in the number of detected genes with and without including introns. X-axis: technology (scRNA-seq or Multiome), y-axis: average number of detected genes, color: intron_inclusion. Link paired data with a line. Repeat the same per cell.
- Intersection of the number of detected genes per intron_inclusion and donor.
- Characterize genes that are only detected after including introns. What is their length? Are they protein-coding or long non-coding RNAs? Are they highly variable? Do they have annotated functions (i.e. GO)?

In the next notebook, we will assess the mixability between scRNA-seq and multiome with and without intron inclusion.

# Pre-processing

## Load packages

```{r}
library(Seurat)
library(tidyverse)
library(ggpubr)
library(here)
library(glue)
library(DT)
library(UpSetR)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(patchwork)
```

## Read data

```{r}
# Read metadata
donor_metadata <- read_csv(
  here("data/tonsil_atlas_donor_metadata.csv"),
  col_names = TRUE
)
DT::datatable(donor_metadata)
sequencing_metadata_rna <- read_csv(
  here("scRNA-seq/1-cellranger_mapping/data/tonsil_atlas_metadata.csv"),
  col_names = TRUE
)
sequencing_metadata_rna$assay <- "scRNA-seq"
sequencing_metadata_multiome <- read_csv(
  here("multiome/1-cellranger_mapping/data/tonsil_atlas_metadata_multiome.csv"),
  col_names = TRUE
)
sequencing_metadata_multiome$assay <- "multiome"
sequencing_metadata <- bind_rows(sequencing_metadata_rna, sequencing_metadata_multiome)
rm(sequencing_metadata_rna, sequencing_metadata_multiome)
selected_gems <- c("d8kwy76j_7blj9otf", "umt51kfr_p8ei65ms", "v8g80gtx_ps9bamz7",
                   "z3of7uaq_mzbhy4tt", "admae8w2_89i88tvv", "co7dzuup_xuczw9vc",
                   "pd9avu0k_kf9ft6kk", "ulx1v6sz_8a2nvf1c")
sequencing_metadata <- sequencing_metadata[sequencing_metadata$gem_id %in% selected_gems, ]
DT::datatable(sequencing_metadata)


# Read expression matrices scRNA-seq
gems_rna <- unique(sequencing_metadata$gem_id[sequencing_metadata$assay == "scRNA-seq"])
gems_multiome <- unique(sequencing_metadata$gem_id[sequencing_metadata$assay == "multiome"])
path_cellranger <- here("scRNA-seq/6-revision/intronic_reads_inclusion/cellranger_mapping")
counts_rna <- list("with_introns" = NULL, "without_introns" = NULL)
for (i in names(counts_rna)) {
  counts <- map(gems_rna, \(x) {
    directories <- list.dirs(glue("{path_cellranger}/scRNA-seq/{i}"), recursive = FALSE)
    directory <- str_subset(directories, x)
    path_to_mat <- glue("{directory}/outs/filtered_feature_bc_matrix")
    print(path_to_mat)
    if (dir.exists(path_to_mat)) {
      counts <- Read10X(path_to_mat)
      return(counts$`Gene Expression`)
    } else{
      return(NULL)
    }
  })
  names(counts) <- gems_rna
  counts_rna[[i]] <- counts
}
rm(counts)
gc()


# Read expression matrices multiome
counts_multiome <- list("with_introns" = NULL, "without_introns" = NULL)
for (i in names(counts_multiome)) {
  counts <- map(gems_multiome, \(x) {
    directories <- list.dirs(glue("{path_cellranger}/multiome/{i}"), recursive = FALSE)
    directory <- str_subset(directories, x)
    path_to_mat <- glue("{directory}/outs/filtered_feature_bc_matrix")
    print(path_to_mat)
    if (dir.exists(path_to_mat)) {
      counts <- Read10X(path_to_mat)
      return(counts$`Gene Expression`)
    } else{
      return(NULL)
    }
  })
  names(counts) <- gems_multiome
  counts_multiome[[i]] <- counts
}
rm(counts)
gc()
```


## Create Seurat objects

```{r}
# RNA
seurat_l_rna <- list("with_introns" = NULL, "without_introns" = NULL)
for (i in names(counts_rna)) {
  counts <- counts_rna[[i]]
  l <- map2(counts, names(counts), \(mat, x) {
    if (!is.null(mat)) {
      seurat_obj <- CreateSeuratObject(counts = mat)
      seurat_obj$gem_id <- x
      seurat_obj$assay <- "scRNA-seq"
      seurat_obj$condition <- i
      seurat_obj <- RenameCells(seurat_obj, add.cell.id = x)
      seurat_obj
    } else{
      NULL
    }
  })
  names(l) <- names(counts)
  seurat_l_rna[[i]] <- l
}
rm(counts, counts_rna, l)
gc() 


# Multiome
seurat_l_multiome <- list("with_introns" = NULL, "without_introns" = NULL)
for (i in names(counts_multiome)) {
  counts <- counts_multiome[[i]]
  l <- map2(counts, names(counts), \(mat, x) {
    if (!is.null(mat)) {
      seurat_obj <- CreateSeuratObject(counts = mat)
      seurat_obj$gem_id <- x
      seurat_obj$assay <- "multiome"
      seurat_obj$condition <- i
      seurat_obj <- RenameCells(seurat_obj, add.cell.id = x)
      seurat_obj
    } else{
      NULL
    }
  })
  names(l) <- names(counts)
  seurat_l_multiome[[i]] <- l
}
rm(counts, counts_multiome, l)
gc() 
```


## Merge

```{r}
# RNA
seurat_rna_introns <- merge(
  x = seurat_l_rna$with_introns$d8kwy76j_7blj9otf,
  y = list(
    seurat_l_rna$with_introns$umt51kfr_p8ei65ms,
    seurat_l_rna$with_introns$v8g80gtx_ps9bamz7,
    seurat_l_rna$with_introns$z3of7uaq_mzbhy4tt
  )
)
seurat_l_rna$with_introns <- NULL
seurat_rna_no_introns <- merge(
  x = seurat_l_rna$without_introns$d8kwy76j_7blj9otf,
  y = list(
    seurat_l_rna$without_introns$umt51kfr_p8ei65ms,
    seurat_l_rna$without_introns$v8g80gtx_ps9bamz7,
    seurat_l_rna$without_introns$z3of7uaq_mzbhy4tt
  )
)
rm(seurat_l_rna)
gc()


# Multiome
seurat_multiome_introns <- merge(
  x = seurat_l_multiome$with_introns$ulx1v6sz_8a2nvf1c,
  y = list(
    seurat_l_multiome$with_introns$co7dzuup_xuczw9vc,
    seurat_l_multiome$with_introns$pd9avu0k_kf9ft6kk,
    seurat_l_multiome$with_introns$admae8w2_89i88tvv
  )
)
seurat_l_multiome$with_introns <- NULL
seurat_multiome_no_introns <- merge(
  x = seurat_l_multiome$without_introns$ulx1v6sz_8a2nvf1c,
  y = list(
    seurat_l_multiome$without_introns$co7dzuup_xuczw9vc,
    seurat_l_multiome$without_introns$pd9avu0k_kf9ft6kk,
    seurat_l_multiome$without_introns$admae8w2_89i88tvv
  )
)
rm(seurat_l_multiome)
gc()
```


# Filter cells

```{r}
all_cells <- read_delim(
  here("data/raw_data_figures/figure_1_all_technologies_umaps.csv"),
  delim = ";",
  col_names = TRUE
)
all_cells$gem_id <- str_sub(all_cells$barcode, 1, 17)
all_cells <- all_cells[all_cells$gem_id %in% c(gems_rna, gems_multiome), ]
seurat_rna_introns <- seurat_rna_introns[, colnames(seurat_rna_introns) %in% all_cells$barcode]
seurat_rna_no_introns <- seurat_rna_no_introns[, colnames(seurat_rna_no_introns) %in% all_cells$barcode]
seurat_multiome_introns <- seurat_multiome_introns[, colnames(seurat_multiome_introns) %in% all_cells$barcode]
seurat_multiome_no_introns <- seurat_multiome_no_introns[, colnames(seurat_multiome_no_introns) %in% all_cells$barcode]
```


# Filter genes

We apply the same filter to the minimum number of cells expressing a gene as we applied in the original analysis.

```{r}
seurat_list <- list(
  "rna_introns" = seurat_rna_introns,
  "rna_no_introns" = seurat_rna_no_introns,
  "multiome_introns" = seurat_multiome_introns,
  "multiome_no_introns" = seurat_multiome_no_introns
)
rm(seurat_rna_introns, seurat_rna_no_introns,
   seurat_multiome_introns,seurat_multiome_no_introns)
gc()
seurat_list <- map(seurat_list, \(seurat_obj) {
  n_cells <- Matrix::rowSums(seurat_obj[["RNA"]]@counts > 0)
  selected_features <- names(n_cells)[n_cells > 5]
  seurat_obj <- subset(seurat_obj, features = selected_features)
  seurat_obj
})
gc()
dir.create(here("scRNA-seq/results/R_objects/intronic_read_mapping"))
saveRDS(
  seurat_list,
  here("scRNA-seq/results/R_objects/intronic_read_mapping/1.seurat_list_rna_multiome_with_and_without_introns.rds")
)
```


# Total number of detected by sample 

```{r}
total_genes_dfs <- map(seurat_list, \(seurat_obj) {
  gems <- unique(seurat_obj$gem_id)
  total_genes <- map_int(gems, \(x) {
    seurat_obj_sub <- seurat_obj[, seurat_obj$gem_id == x]
    gene_counts <- Matrix::rowSums(seurat_obj_sub[["RNA"]]@counts > 0)
    total <- sum(gene_counts > 5)
    total
  })
  df <- data.frame(gem_id = gems, total_genes_detected = total_genes)
  df
})
total_genes_df <- bind_rows(total_genes_dfs, .id = "group")
total_genes_df$assay <- ifelse(
  str_detect(total_genes_df$group, "rna"),
  "scRNA-seq",
  "multiome"
)
total_genes_df$condition <- ifelse(
  str_detect(total_genes_df$group, "no_introns"),
  "without introns",
  "with introns"
)
selected <- sequencing_metadata$gem_id %in% c(gems_rna, gems_multiome)
sequencing_metadata <- sequencing_metadata[selected, c("gem_id", "donor_id")]
sequencing_metadata <- sequencing_metadata[!duplicated(sequencing_metadata), ]
total_genes_df <- left_join(total_genes_df, sequencing_metadata, by = "gem_id")
total_genes_gg <- total_genes_df %>%
  mutate(
    assay = factor(assay, levels = c("scRNA-seq", "multiome")),
    condition = factor(condition, levels = c("without introns", "with introns")),
    donor_id = factor(donor_id, levels = c("BCLL-8-T", "BCLL-9-T", "BCLL-14-T", "BCLL-15-T"))
  ) %>% 
  ggplot(aes(condition, total_genes_detected, color = assay)) +
    geom_line(aes(group = donor_id), color = "gray35") +
    geom_point() +
    facet_wrap(~assay) +
    stat_compare_means(
      comparisons = list(c("with introns", "without introns")),
      method = "t.test",
      paired = TRUE,
      label =  "p.signif"
    ) +
    scale_color_manual(values = c("#d52f55", "#4aafdd")) +
    labs(y = "Total number of detected\ngenes (per sample)") +
    scale_y_continuous(limits = c(15000, 25000)) +
    theme_bw() +
    theme(
      axis.title.x = element_blank(),
      legend.position = "none",
      axis.text.x = element_text(size = 7, angle = 45, vjust = 1, hjust = 1),
      axis.text.y = element_text(size = 7),
      axis.title.y = element_text(size = 8),
      strip.text = element_text(size = 7))
total_genes_gg
compare_means(total_genes_detected ~ condition, data = total_genes_df, 
              group.by = "assay", method = "t.test", paired = TRUE)
ggsave(
  plot = total_genes_gg,
  filename = here("results/paper/figures/revision_intron_inclusion_total_genes_detected.pdf"),
  height = 8,
  width = 10,
  units = "cm"
)
```

Per cell:

```{r}
metadata_dfs <- map(seurat_list, \(seurat_obj) {
  seurat_obj$barcode <- colnames(seurat_obj)
  df <- seurat_obj@meta.data[, c("barcode", "gem_id", "condition", "assay", "nFeature_RNA")]
  df
})
metadata_df <- bind_rows(metadata_dfs)
metadata_df <- metadata_df %>%
  mutate(
    assay = factor(assay, levels = c("scRNA-seq", "multiome")),
    condition = factor(condition, levels = c("without_introns", "with_introns")),
  )
levels(metadata_df$condition) <- c("without introns", "with introns")
unique_barcodes <- table(str_sub(metadata_df$barcode, 1, 34)) == 1
unique_barcodes <- names(unique_barcodes)[unique_barcodes]
pattern <- str_c(unique_barcodes, collapse = "|")
metadata_df <- metadata_df[str_detect(metadata_df$barcode, pattern, negate = TRUE), ]
number_genes_per_cell <- metadata_df %>%
  ggplot(aes(condition, nFeature_RNA, color = assay)) +
    geom_violin(fill = NA) +
    geom_boxplot(width = 0.15, outlier.size = 0.01) +
    facet_wrap(~assay) +
    stat_compare_means(
      comparisons = list(c("with introns", "without introns")),
      method = "t.test",
      paired = TRUE,
      label =  "p.signif"
    ) +
    scale_color_manual(values = c("#d52f55", "#4aafdd")) +
    scale_y_continuous(limits = c(0, 17500)) +
    ylab("Number of detected\ngenes (per cell)") +
    theme_bw() +
    theme(
      axis.title.x = element_blank(),
      legend.position = "none",
      axis.title.y = element_text(size = 8),
      strip.text = element_text(size = 7),
      axis.text.y = element_text(size = 7),
      axis.text.x = element_text(size = 7, angle = 45, vjust = 1, hjust = 1)
    )
number_genes_per_cell
ggsave(
  plot = number_genes_per_cell,
  filename = here("results/paper/figures/revision_intron_inclusion_number_genes_detected_per_cell.pdf"),
  height = 8,
  width = 10,
  units = "cm"
)
``` 


# Intersection of detected genes

```{r}
detected_genes <- map(seurat_list, rownames)
names(detected_genes) <- c("scRNA-seq\nwith introns", "scRNA-seq\nwithout introns",
                           "multiome\nwith introns", "multiome\nwithout introns")
in2mm <- 25.4
path_to_save_upset_plot <- here("results/paper/figures/revision_intron_inclusion_upset_plot_intersection_detected_genes.pdf")
pdf(path_to_save_upset_plot, width = (115 / in2mm), height = (105 / in2mm))
upset(fromList(detected_genes), order.by = "freq", nintersects = 8, mb.ratio = c(0.55, 0.45))
dev.off()
```


# Length newly detected genes

```{r}
# txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
gtf <- read_tsv(here("data/cellranger_subsetted_reference.gtf"), col_names = FALSE)
gtf <- separate(gtf, col = "X9", sep = ";", into = str_c("X", 10:17))
gtf_sub <- gtf[, c("X3", "X4", "X5", "X12", "X13")]
gtf_sub$X12 <- str_remove(gtf_sub$X12, ' gene_type \"')
gtf_sub$X12 <- str_remove(gtf_sub$X12, '\"')
gtf_sub$X13 <- str_remove(gtf_sub$X13, ' gene_name \"')
gtf_sub$X13 <- str_remove(gtf_sub$X13, '\"')
colnames(gtf_sub) <- c("sequence", "start",  "end", "gene_type", "gene")
gtf_sub$length <- gtf_sub$end - gtf_sub$start
gtf_sub <- gtf_sub[!duplicated(gtf_sub$gene), ]
gtf_sub <- as.data.frame(gtf_sub)
rownames(gtf_sub) <- gtf_sub$gene
intersection_genes <- intersect(
  x = detected_genes$`scRNA-seq\nwith introns`,
  y = detected_genes$`scRNA-seq\nwithout introns`
)
non_intersection_genes <- detected_genes[[1]][!(detected_genes[[1]] %in% intersection_genes)]
gtf_intersecting <- gtf_sub[intersection_genes, ]
gtf_non_intersecting <- gtf_sub[non_intersection_genes, ]
gtf_intersecting$group <- "common genes"
gtf_non_intersecting$group <- "only captured\nwith intron inclusion"
gtfs <- list(gtf_intersecting, gtf_non_intersecting)
gtfs <- map(gtfs, \(df) {
  df <- df %>%
    mutate(gene_type = case_when(
      gene_type == "protein_coding" ~ "protein_coding",
      gene_type == "lncRNA" ~ "lncRNA",
      TRUE ~ "other"
    )) %>%
    mutate(gene_type = factor(gene_type, c("protein_coding", "lncRNA", "other")))
  levels(df$gene_type) <- c("protein coding", "lncRNA", "other")
  df
})
pie_charts <- map(gtfs, \(df) {
  p <- df %>%
    group_by(gene_type) %>%
    summarize(n_genes = n()) %>%
    mutate(total_genes = sum(n_genes), pct_genes = n_genes / total_genes * 100) %>%
    ggplot(aes(x = "", y = pct_genes, fill = gene_type)) +
    geom_bar(stat = "identity", width = 1) +
    coord_polar("y", start = 0, direction = -1) +
    scale_fill_manual(values = c("#552328", "#22446a", "#eb8259")) +
    theme_classic() +
    theme(
      axis.line = element_blank(),
      axis.ticks = element_blank(),
      axis.title = element_blank(),
      axis.text = element_blank(),
      legend.title = element_blank(),
      legend.text = element_text(size = 7))
  p
})
pie_charts[[1]] <- pie_charts[[1]] + 
  theme(legend.position = "none") +
  labs(title = "Common genes", subtitle = glue("n = {length(intersection_genes)}")) +
  theme(plot.title = element_text(size = 9, hjust = 0.5),
        plot.subtitle = element_text(size = 8))
pie_charts[[2]] <- pie_charts[[2]] + 
  labs(title = "only captured\nwith intron inclusion", subtitle = glue("n = {length(non_intersection_genes)}")) +
  theme(plot.title = element_text(size = 9, hjust = 0.5),
        plot.subtitle = element_text(size = 8))
pie_charts_arranged <- pie_charts[[1]] | pie_charts[[2]]
pie_charts_arranged
ggsave(
  plot = pie_charts_arranged,
  filename = here("results/paper/figures/revision_intron_inclusion_pie_charts_gene_types.pdf"),
  height = 6,
  width = 10,
  units = "cm"
)
```



# Session Information

```{r}
sessionInfo()
```

