---
title: "Calling Peaks"
author: "Paula Soler-Vila"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE, out.width = "100%", fig.align = "center",
  message = FALSE, warning = FALSE
)
options(width = 1200)
```

# Introduction
Looking at the scATAC-seq coverage plots, it's possible to vislumbrate that, as a consequence of the merging process, some peaks show a considerable increase in size. This can create a problem for certain analysis such as motif enrichment analysis. In order to solve it, we decided to do a new round of peak calling using MACS2 for each group of cells annotated at level 1, helping to identify specifics peaks for rarer populations.  

# Pre-processing

## Load packages

```{r libraries} 
library(Seurat)
library(Signac)
library(reshape)
library(ggplot2)
library(GenomicRanges)
library(EnsDb.Hsapiens.v86)
library(stringr)
library(plotly)
```

## Parameters

```{r parameters}
cell_type = "GCBC"

# Paths
path_to_obj <- paste0(
  here::here("scATAC-seq/results/R_objects/level_4/"),
  cell_type,
  "/",
  cell_type,
  "_annotated_level_4.rds",
  sep = ""
)

path_to_save <- paste0(
  here::here("scATAC-seq/results/R_objects/level_4/"),
  cell_type,
  "/",
  cell_type,
  "_annotated_peak_calling_level_4.rds",
  sep = ""
)

color_palette <-  c("#1CFFCE", "#90AD1C", "#C075A6", "#85660D", 
                    "#5A5156", "#AA0DFE", "#F8A19F", "#F7E1A0", 
                    "#1C8356", "#FEAF16", "#822E1C", "#C4451C", 
                    "#1CBE4F", "#325A9B", "#F6222E", "#FE00FA", 
                    "#FBE426", "#16FF32", "black", "#3283FE",
                    "#B00068", "#DEA0FD", "#B10DA1", "#E4E1E3",
                    "#90AD1C", "#FE00FA", "#85660D", "#3B00FB",
                    "#822E1C", "coral2", "#1CFFCE", "#1CBE4F",
                    "#3283FE", "#FBE426", "#F7E1A0", "#325A9B",   
                    "#2ED9FF", "#B5EFB5", "#5A5156", "#DEA0FD",
                    "#FEAF16", "#683B79", "#B10DA1", "#1C7F93", 
                    "#F8A19F", "dark orange", "#FEAF16", "#FBE426",  
                    "Brown")
```

## Functions

```{r functions}
peaks_quantification <- function(seurat_filtered, new_peaks){ 
    
    annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
    #seqlevelsStyle(annotation) <- "UCSC"
    ucsc.levels <- str_replace(string=paste("chr",seqlevels(annotation),sep=""), pattern="chrMT", replacement="chrM")
    seqlevels(annotation) <- ucsc.levels
    genome(annotation) <- "hg38"

    macs2_counts <- FeatureMatrix(
    fragments = Fragments(seurat_filtered),
    features = new_peaks,
    cells = colnames(seurat_filtered))
    
    seurat_filtered[["peaks_redefined"]] <- CreateChromatinAssay(
      counts = macs2_counts, 
     # genome = "hg38",
      fragments = Fragments(seurat_filtered),
      annotation = annotation)
    
    return(seurat_filtered)
}
```

# Load data

```{r}
seurat <- readRDS(path_to_obj)

Idents(seurat) <- seurat$level_5
```

```{r}
DimPlot(seurat,
  group.by = "level_5",
  cols = color_palette,
  pt.size = 0.5)
```

# 3D representation of the clusters
```{r}
# Re-run UMAPs that you have accurate calculations for all UMAP(s)
yourseuratobject <- RunUMAP(reduction = 'harmony', seurat,
                            dims = 2:30,  
                            n.components = 3L)

# Extract tSNE information from Seurat Object
umap_1 <- yourseuratobject[["umap"]]@cell.embeddings[,1]
umap_2 <- yourseuratobject[["umap"]]@cell.embeddings[,2]
umap_3 <- yourseuratobject[["umap"]]@cell.embeddings[,3]

# Prepare a dataframe for cell plotting
plot.data <- FetchData(object = yourseuratobject, vars = c("UMAP_1", "UMAP_2", "UMAP_3", "level_5"))

# Make a column of row name identities (these will be your cell/barcode names)
plot.data$label <- paste(rownames(plot.data))

fig <- plot_ly(data = plot.data, 
        x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
        color = ~level_5, 
        colors = c("gray10","gray20","gray30",
                   "gray40","gray50",
                   "red","orange","pink","gray60"),
        type = "scatter3d", 
        mode = "markers", 
        marker = list(size = 3, width=2), # controls size of points
        text=~label, #This is that extra column we made earlier for which we will use for cell ID
        hoverinfo="text") #When you visualize your plotly object, hovering your mouse pointer over a point shows cell names
fig
```


# Coverage plot
```{r}
CoveragePlot(seurat, 
             region = "chr8-127736231-127742951")
```


# New round of Peak Calling using MACS2 on each annotated cell type at level 5.

```{r}
peaks_level5 <- CallPeaks(
  object = seurat,
  group.by = "level_5",
  macs2.path = "/Users/pauli/opt/anaconda3/envs/Tonsil_atlas/bin/macs2")
```

```{r}
# remove peaks on nonstandard chromosomes and in genomic blacklist regions
peaks_level5 <- keepStandardChromosomes(peaks_level5, pruning.mode = "coarse")
peaks_level5 <- subsetByOverlaps(x = peaks_level5, 
                         ranges = blacklist_hg38_unified, 
                         invert = TRUE)
```

# Quantification of the new peak counts in the entire dataset

```{r}
seurat <- peaks_quantification(seurat, peaks_level5)
```

```{r}
# remove the initial peaks annotation
DefaultAssay(seurat) <- 'peaks_redefined'
seurat[['peaks_macs']] <- NULL
seurat

saveRDS(seurat,path_to_save)
```

# Coverage plot
```{r}
CoveragePlot(seurat, 
             ranges = peaks_level5,
             region = "chr8-127736231-127742951")
```

# Session Information
```{r session}
sessionInfo()
```
