---
author: "Marc Elosua-Bayes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
title: "GC T cells SC mapping"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE, cache = FALSE,
                      cache.lazy = FALSE)
options(width = 1200)
set.seed(123)
```

## Introduction
In this script we are going to map T-cell subtypes onto the Visium slides.

## Libraries
```{r}
library(Seurat)
library(ggpubr)
library(cowplot)
library(dplyr)
library(ggplot2)
library(stringr)
library(readr)
library(SPOTlight)
```

## Setting parameters
Loading necessary paths and parameters
```{r}
set.seed(123)
source(here::here("misc/paths.R"))
source(here::here("utils/bin.R"))

"{gct}/{plt_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = ,
             showWarnings = FALSE,
             recursive = TRUE)

"{gct}/{robj_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = ,
             showWarnings = FALSE,
             recursive = TRUE)

gc_vec <- c(
    "DZ G2Mphase", "DZ Sphase", "DZ_LZ transition", "DZ non proliferative",
    "LZ", "prePC", "LZ-DZ-re-entry", "MBC")
```

Extract sample id and get Donor ID
```{r}
# sample_id <- params$sample_id
sample_id <- "esvq52_nluss5"
donor_id <- id_sp_df[id_sp_df$gem_id == sample_id, ] %>% dplyr::pull(donor_id)
```

## Load data
The spatial data comes from the script *03-clustering/03-clustering_integration.Rmd* while the sc data can be found in Ramon's scRNAseq analysis: **/scratch/devel/rmassoni/tonsil_atlas_private/2-DOWNSTREAM_PROCESSING/results/R_objects/processed_seurat_objects/processed_seurat_objects/tonsil_integrated_with_harmony_scrublet_annotated.rds**. \
```{r}
sp_obj <- "misc/{robj_dir}/20220215_tonsil_atlas_spatial_seurat_obj.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)

# Single cell data
sc_obj <- "misc/{robj_dir}/20220215_tonsil_atlas_rna_seurat_obj.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)

# GC data
sc_gc <- "{gct}/{robj_dir}/GCBC_seu_obj_level_5_eta.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

## Analysis

### Join GC to full object
```{r}
# Precursor MBCs are NA in the ATAC annotation
table(
    sc_gc$names_level_5_clusters_eta_atac,
    sc_gc$names_level_5_clusters_eta,
    useNA = "ifany")

gc_meta <- sc_gc@meta.data %>%
    tibble::rownames_to_column("bc") %>%
    dplyr::select(bc, names_level_5_clusters_eta_atac) %>%
    dplyr::rename(gc_annot = names_level_5_clusters_eta_atac) %>%
    dplyr::mutate(gc_annot = if_else(is.na(gc_annot), "MBC", gc_annot))

table(gc_meta$gc_annot)

sc_obj@meta.data <- sc_obj@meta.data %>%
    tibble::rownames_to_column("bc") %>%
    dplyr::left_join(gc_meta, by = "bc") %>%
    dplyr::mutate(
        annotation_20220215_mod = if_else(
            is.na(gc_annot), annotation_20220215, gc_annot)
    ) %>%
    tibble::column_to_rownames("bc")
```

### Metadata
We are going to use SPOTlight to deconvolute the spatial spots and map the cell types.

We will start by adding the level5 definitive metadata labels to the full object.
```{r}
sc_obj@meta.data <- sc_obj@meta.data %>%
  dplyr::mutate(
    # Annotation for SPOTlight
    annotation_deconv = dplyr::if_else(
      annotation_20220215_mod %in% gc_vec,
      as.character(annotation_20220215_mod), as.character(annotation_level_1)),
    # Annotation for SPOTlight with condensed names
    # annotation_deconv = dplyr::case_when(
    #   annotation_deconv %in% c("Naive CD8 T", "SCM CD8 T", "CM CD8 T") ~ "Naive-CM CD8 T",
    #   annotation_deconv %in% c("CXCR6+ RM CD8 T", "RM CD8 T") ~ "CXCR6pn RM CD8 T",
    #   TRUE ~ annotation_deconv)
    )

# Only keep cell types found within the follicles
# Remove GCBC since all the GCBC cells are already in lvl5
table(sc_obj@meta.data$annotation_deconv, sc_obj@meta.data$annotation_level_1)
sc_obj <- sc_obj[, sc_obj$annotation_deconv %in% c(gc_vec, "FDC", "CD4_T", "NBC_MBC")]

table(sc_obj@meta.data$annotation_deconv)

# level 5 GC T Cells
exist <- gc_vec %in% unique(sc_obj$annotation_deconv)
names(exist) <- gc_vec
exist
table(exist)
lvl5_gc_obj <- sc_obj[, sc_obj$annotation_deconv %in% gc_vec]
```

### Marker calculation
Since the HVG selection was carried out with the entire dataset we are going to do normalization again within this subset
```{r}
sc_obj <- Seurat::NormalizeData(
  sc_obj,
  normalization.method = "LogNormalize",
  scale.factor = 10000,
  verbose = FALSE) %>%
  Seurat::FindVariableFeatures(., nfeatures = 3000, verbose = FALSE)
```

1st we need to determine the marker genes for each high-level cluster
```{r}
markers_lvl1_out <- "{gct}/{robj_dir}/markers_lvl1_GC.RDS" %>% 
  glue::glue() %>% 
  here::here()
# file.remove(markers_lvl1_out)
if (file.exists(markers_lvl1_out)) {
  
  markers_lvl1 <- readRDS(file = markers_lvl1_out)
  
} else {
  #### Extract marker genes from each cluster ####
  Seurat::Idents(object = sc_obj) <- sc_obj@meta.data[, "annotation_level_1"]
  markers_lvl1 <- Seurat::FindAllMarkers(
    object = sc_obj, 
    assay = "RNA",
    slot = "data",
    verbose = TRUE, 
    only.pos = TRUE,
    max.cells.per.ident = 200)
  
  saveRDS(object = markers_lvl1, file = markers_lvl1_out)
}
DT::datatable(markers_lvl1, filter = "top")
```

Now find the markers for GC T cells level 5 only between GC T cells
```{r}
markers_lvl5_out <- "{gct}/{robj_dir}/markers_lvl5_gc.RDS" %>% 
  glue::glue() %>% 
  here::here()
# file.remove(markers_lvl5_out)

if (file.exists(markers_lvl5_out)) {
  
  markers_lvl5 <- readRDS(file = markers_lvl5_out)
  
} else {

  #### Extract marker genes from each cluster ####
  Seurat::Idents(object = lvl5_gc_obj) <- lvl5_gc_obj@meta.data[, "annotation_deconv"]
  
  markers_lvl5 <- Seurat::FindAllMarkers(
    object = lvl5_gc_obj, 
    assay = "RNA",
    slot = "data",
    verbose = TRUE, 
    only.pos = TRUE,
    max.cells.per.ident = 200) 
  
  saveRDS(object = markers_lvl5, file = markers_lvl5_out)
}

DT::datatable(markers_lvl5, filter = "top")
```

Join markers from both runs. Since we still want to keep the CD4 T cells markers in the lower level markers we will add them to each one.
```{r}
# Keep markers for major cell types
markers_lvl1_sub <- markers_lvl1 %>%
    dplyr::filter( ( avg_log2FC > 1 | pct.1 - pct.2 > 0.75) &
                   p_val_adj < 0.01 & pct.1 > 0.3 ) %>%
    dplyr::filter(cluster != "GCBC")
```

Add general GC T cell markers to the specific cell types
```{r}
# markers_lvl5_update <- lapply(as.character(unique(markers_lvl5$cluster)),
#   function(i) {
markers_lvl5_update <- markers_lvl5 %>%
  # dplyr::filter(cluster == i) %>%
  # dplyr::bind_rows(markers_lvl1 %>%
  #     filter(cluster == "GCBC" & avg_log2FC > 1.5)) %>%
  # dplyr::mutate(cluster = i) %>%
  dplyr::filter(avg_log2FC > 0.75 | pct.1 - pct.2 > 0.3)
# }) %>% bind_rows()
```

Combine level-1 and level-5 markers
```{r}
all_markers <- dplyr::bind_rows(markers_lvl1_sub, markers_lvl5_update)
```

Since we want to limit the batch effect we will try to select all the cells from the same cell type from the fewest batches as possible.
```{r}
# subset most representative sample(s)
ns <- with(sc_obj@meta.data, table(gem_id, annotation_deconv))
ns <- as.matrix(unclass(ns))
m <- 100 # Max cells per cell type
# Extract cell barcodes
meta <- sc_obj@meta.data

id <- lapply(colnames(ns), function(nm) {
    x <- ns[, nm]
    # Initialize variables
    n <- 0    # N of cells
    s <- c()  # Gem IDs
    b <- c()  # Cell barcodes
    while (n < m & length(x) > 0) {
        # select gem id with the most cells
        i <- which.max(x)
        # Add gem id to vector s
        s <- c(s, names(i))
        # Add number of cells per cell type to n
        n <- n + x[i]
        # Remove gem id from x to move on to the next
        x <- x[-i]
        # extract barcode
        barcode <- rownames(meta[meta[, "gem_id"] == names(i) &
                            meta[, "annotation_deconv"] == nm, ])
        # make sure it adds up to m
        print(barcode)
        if (length(b) + length(barcode) > m) {
          barcode <- sample(x = barcode, size = m - length(b), replace = FALSE)
        }
        b <- c(b, barcode)  # Cell barcodes
        # print(b)

    }
    return(b)
})

sc_sub <- sc_obj[, unlist(id)]
table(sc_sub$annotation_deconv)
```

### SPOTlight deconvolution
2nd run deconvolution
```{r}
Seurat::Idents(object = sc_sub) <- as.character(sc_sub@meta.data$annotation_deconv)
# sc_obj@meta.data <- sc_obj@meta.data %>% tibble::column_to_rownames("cell_barcode")

spotlight_ls <- spotlight_deconvolution(
  se_sc = sc_sub,
  counts_spatial = sp_obj@assays$Spatial@counts,
  clust_vr = "annotation_deconv",
  cluster_markers = all_markers,
  cl_n = 100,
  hvg = 3000,
  ntop = NULL,
  transf = "uv",
  method = "nsNMF",
  min_cont = 0)
```

## Save data
```{r}
decon_fn <- "{gct}/{robj_dir}/spotlight_ls_gc.rds" %>%
  glue::glue() %>% 
  here::here()
# if (! file.exists(decon_fn) ) {
saveRDS(
  object = spotlight_ls,
  file = decon_fn)
# } else {
#   spotlight_ls <- readRDS(file = here::here(glue::glue("{gct}/{robj_dir}/spotlight_ls_cd4.rds")))
# }

```

3rd Check Topic profiles
```{r fig.height=8, fig.width=8}
nmf_mod_ls <- spotlight_ls[[1]]
nmf_mod <- nmf_mod_ls[[1]]

h <- NMF::coef(nmf_mod)
rownames(h) <- paste("Topic", 1:nrow(h), sep = "_")
topic_profile_plts <- SPOTlight::dot_plot_profiles_fun(
  h = h,
  train_cell_clust = nmf_mod_ls[[2]])

topic_profile_plts[[2]] +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90), 
                 axis.text = ggplot2::element_text(size = 12))
```

## Visualize deconvolution results
```{r}
decon_mtrx <- spotlight_ls[[2]]
decon_mtrx <- decon_mtrx[, colnames(decon_mtrx) != "res_ss"]

# Set as 0 cell types predicted to be under 1 % of the spot
# decon_mtrx[decon_mtrx < 0.03] <- 0
```

Change column names
```{r}
nm_df <- data.frame(
  mod_nm = stringr::str_replace_all(string = unique(all_markers$cluster),
                                    pattern = "[[:punct:]]|\\s|\\+" ,
                                    replacement = "."),
  plt_nm = unique(all_markers$cluster)
  )

# Save nm_df
"{gct}/{robj_dir}/gc_nm_df.rds" %>%
  glue::glue() %>% 
  here::here() %>%
  saveRDS(
    object = nm_df,
    file = .)


new_cn <- data.frame(mod_nm = colnames(decon_mtrx)) %>%
  dplyr::left_join(nm_df, by = "mod_nm") %>%
  # Central.Mem.PASK. fives some trouble because it only changes between + an -
  # negative goes first and distinct solves it automatically
  dplyr::distinct() %>%
  dplyr::pull(plt_nm)

colnames(decon_mtrx) <- new_cn
```

We are going to add the deconvolution to the Seurat object.
```{r}
metadata <- cbind(sp_obj@meta.data, decon_mtrx)
sp_obj@meta.data <- metadata
```

Look at cells topic profile
```{r}
basis_spotlight <- data.frame(NMF::basis(spotlight_ls[[1]][[1]]))

train_labs <- spotlight_ls[[1]][[2]]
colnames(basis_spotlight) <- unique(stringr::str_wrap(train_labs, width = 30))

basis_spotlight[basis_spotlight < 0.0000001] <- 0

DT::datatable(basis_spotlight, filter = "top")
```

#### Cell type location
Look at the location of each cell type in each slice separately
```{r fig.height=25, fig.width=25}
# Iterate over cell types
ct <- colnames(decon_mtrx)
sp_sub <- sp_obj[, sp_obj@meta.data$gem_id == sample_id]
sp_sub@images <- sp_sub@images[Seurat::Images(sp_sub) == sample_id]

ct_plt_ls <- lapply(ct, function(i) {
  tmp_plt <- Seurat::SpatialFeaturePlot(object = sp_sub,
                                        features = i,
                                        alpha = c(0, 1)) +
    ggplot2::scale_fill_gradientn(
      colors = heat.colors(10, rev = TRUE)) +
    ggplot2::scale_alpha(range = c(0, 1)) +
    ggplot2::labs(title = stringr::str_wrap(string = i,
                                   width = 25),
         fill = "") +
    ggplot2::theme(
      plot.title = ggplot2::element_text(
        hjust = 0.5,
        size = 20,
        face = "bold"))
  
  if (sum(sp_sub@meta.data[, i]) == 0) {
    tmp_plt <- suppressMessages(tmp_plt + ggplot2::scale_alpha(range = c(0,0)))
  }
  
  return(tmp_plt)
})

(plt_arr <- cowplot::plot_grid(
  plotlist = ct_plt_ls,
  axis = "trbl",
  align = "hv",
  nrow = 5,
  ncol = 5))

"{gct}/{plt_dir}/cell_type_location_gc_{donor_id}.pdf" %>%
  glue::glue() %>%
  here::here() %>%
  cowplot::save_plot(
    filename = .,
    plot = plt_arr,
    base_height = 25,
    base_width = 25)
```

## Session Info
```{r}
sessionInfo()
```
