---
author: "Marc Elosua-Bayes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
params: 
    sample_id: "Default!"
    donor_id: "Default!"
title: "CD4 MAGIC denoising"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# knitr::opts_knit$set(root.dir = "..")
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE)
options(width = 1200)
```

## Introduction
In this Rmarkdown document we are going to use MAGIC to better visualize genes and ease with the annotation when using specific marker genes.
[MAGIC](https://magic.readthedocs.io/en/stable/) was developed by Smita Krishnaswamy's lab to try to fill in the drop out reads in the spots. `MAGIC` is a Markov Affinity-based Graph Imputation of Cells used for denoising high-dimensional data most commonly applied to single-cell RNA sequencing data. MAGIC learns the manifold data, using the resultant graph to smooth the features and restore the structure of the data based on their k-nearest neighbors. 

## Libraries
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
# library(SPATA2)
```

Loading *Rmagic*
```{r}
library(reticulate)

# conda create -n MAGIC python=3.7
# conda install -c bioconda scprep
# conda install matplotlib future tasklogger graphtools scipy pandas Deprecated pygsp pytz python-dateutil six threadpoolctl joblib decorator wrapt cycler kiwisolver pyparsing pillow
# conda install -c anaconda zlib
# ~/anaconda3/envs/MAGIC/bin/pip3 install magic-impute
#
# path_to_python <- "/media/data1/anaconda3/envs/MAGIC"
path_to_python <- "/scratch/groups/hheyn/software/anaconda3/envs/spatial_r/"
# reticulate::use_python(path_to_python, required = TRUE)
reticulate::use_condaenv(path_to_python)
reticulate::py_discover_config(required_module = "magic")
reticulate::py_config()
library(Rmagic)
```

## Parameters
```{r}
set.seed(123)
source(here::here("misc/paths.R"))
source(here::here("utils/bin.R"))

"{cd4}/{plt_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = .,
             showWarnings = FALSE,
             recursive = TRUE)

"{cd4}/{robj_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = .,
             showWarnings = FALSE,
             recursive = TRUE)
```

## Load data
The data used in this Rmarkdown document comes from **03-clustering_integration.Rmd** where the data was integrated.
```{r}
merged_se <- "{clust}/{robj_dir}/integrated_spatial.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

## Analysis

### MAGIC denoising

Run MAGIC to denoise the data. The markers used here come from 2 papers published on Tfh differentiation
The markers denoised and used here come from 2 papers published on Tfh differentiation \
- [T Follicular Helper Cell Biology: A Decade of Discovery and Diseases](https://doi.org/10.1016/j.immuni.2019.04.011) \
- [Bcl6-Mediated Transcriptional Regulation of Follicular Helper T cells (Tfh)](https://doi.org/10.1016/j.it.2021.02.002)

- **CCR7** Gene exrpressed by DC and naive T cells that allows the migration and homing to the lymphoid organs. Also expressed by B cells so that they can migrate towards the T:B cell border. \
- **CD40** Gene expressed by cognate B cells to interact with Tfh and introduce them into the follicle \
- **IL6** Interleukin-6 (IL-6), secreted by murine monocytes and B cells, and the costimulatory molecule ICOSL are early factors involved in Tfh differentiation \
- **ICOSL** Interleukin-6 (IL-6) and the costimulatory molecule ICOSL are early factors involved in Tfh differentiation. Expressed by B cells. \
- **IL2** IL-2 is the most potent inhibitor of Tfh differentiation because it induces Blimp-1.  \
- **CD80** MHC-II Costimulatory molecule expressed by B cells to make the APC more potent \
- **CD86** MHC-II Costimulatory molecule expressed by B cells to make the APC more potent \
- **CXCR5** Receptor of CXCL13 that leads Tfh and B cells towards the follicle structure where it will meet the B cell. \
- **CXCL13** Secreted by FDC ti attract CXCR5 exoression cells. \
- **ICOS** Expressed by Tfh and binds with ICOSL from the B cell to make the MHC-II presentation stronger. ICOS is particularly fascinating, as it functions as both a costimulatory molecule and a migration receptor for Tfh \
- **SELPLG** gene name for *PSGL1* -  migration of Tfh cell into GCs is facilitated by repression of the migration-associated receptor PSGL1 and the chemoattractant receptor Ebi2 \
- **GPR183** gene name for *Ebi2* - migration of Tfh cell into GCs is facilitated by repression of the migration-associated receptor PSGL1 and the chemoattractant receptor Ebi2 \
- **TCF7** gene name for TCF1 - TCF1 and LEF1 are involved in induction of Bcl6, as are STAT3 and STAT1 \
- **LEF1** TCF1 and LEF1 are involved in induction of Bcl6, as are STAT3 and STAT1 \
- **STAT3** TCF1 and LEF1 are involved in induction of Bcl6, as are STAT3 and STAT1 \
- **STAT1** TCF1 and LEF1 are involved in induction of Bcl6, as are STAT3 and STAT1 \
- **CD200** is expressed in GC Tfh \
- **BTLA** is expressed in GC Tfh \
- **PDCD1** is expressed in GC Tfh \
- **IL12** IL-12 is the most potent human cytokine at inducing IL-21-expressing CD4 + T cells, and IL-12 signals through different STATs than IL-6 \
- **IL21** Tfh cells provide IL-21 and CD40L signals required for B cell proliferation and differentiation toward both GC and extrafollicular fates \
- **CD40LG** Tfh cells provide IL-21 and CD40L signals required for B cell proliferation and differentiation toward both GC and extrafollicular fates \
- **IL4** IL-4 expression by Tfh cells is only acquired later, upon full differentiation to GC-Tfh cells \
- **PSGL1 (SELPLG), GPR183** The migration of Tfh cell into GCs is ages, facilitated by repression of the migration-associated receptor PSGL1 and the chemoattractant receptor Ebi2, and changes in the expression of S1P receptors, SLAM family and integrins (Meli et al., 2016; Vinuesa et al., 2016). \

[A Spatially Resolved Dark- versus Light-Zone Microenvironment Signature Subdivides Germinal Center-Related Aggressive B Cell Lymphomas](10.1016/j.isci.2020.101562)
- **Dark Zone**: OAZ1, AICDA, H3, MKI67, POLH
- **Light Zone**: LAG3, ITGB8, PDCD1, TIGIT, BCL2, PECAM1, LY6E, B7-H3 (CD276), HLA-DRB1, PSMB10, TNF, ARG1, HLA-E, STAT1

[A cell atlas of human thymic development defines T cell repertoire formation](https://www.science.org/doi/10.1126/science.aay3224)
- **Non-activated DC**: ITGAX, XCR1

```{r}
"{cd4}/gene_dict.R" %>%
  glue::glue() %>%
  here::here() %>%
  source(file = .)

gene_vec <- gene_vec[gene_vec %in% rownames(merged_se)]
```

Run MAGIC, it is recommended by the developers to run it within sample so we are going to run it separately for each one.
This follows the same principle as why we want to run SCTransform in samples individually, we don't want information leaking from one dataset to another even if they are healthy tonsils that should be homogeneous.
```{r}
magic_ls <- lapply(id_sp_df$gem_id, function(id) {
  data_magic <- Rmagic::magic(
    data = merged_se[, merged_se$gem_id == id],
    assay = "Spatial",
    counts = "data",
    genes = gene_vec,
    knn = 2,
    knn.max = NULL,
    decay = 1,
    t = "auto",
    npca = 100,
    init = NULL,
    t.max = 20,
    knn.dist.method = "euclidean",
    verbose = 1,
    n.jobs = 1,
    seed = 123)
  
  tmp_mtrx <- data_magic@assays$MAGIC_Spatial@data
  data.frame(tmp_mtrx, check.names = FALSE)
})

magic_df <- lapply(magic_ls, function(i) data.frame(i, check.names = FALSE)) %>%
  dplyr::bind_cols()

"{cd4}/{robj_dir}/MAGIC-mtrx.rds" %>%
  glue::glue() %>%
  here::here() %>%
  saveRDS(object = magic_df, file = .)
```

## Session Info
```{r}
sessionInfo()
```
