---
title: "Cellranger Mapping QC (MCL)"
author: "Ramon Massoni-Badosa"
date: "`r format(Sys.time(), '%B %Y')`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction

The objective of this notebook is to perform a basic quality control of the mapping performed with [Cellranger](https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/what-is-cell-ranger-arc). Cell Ranger ARC analyzes data generated by the Chromium Single Cell Multiome ATAC + Gene Expression. The expected values (EV) included in the notebook are obtained from this [technical_note](https://assets.ctfassets.net/an68im79xiti/6uFMwMQwMySzluyIeBZhMl/e5acb853870c59d3f3fc826428cb6d3c/CG000368_TechnicalNote_InterpretingARCWebSummaryFilesforMultiomeATAC_GEX_RevA.pdf)


# Pre-processing

## Load packages

```{r}
library(ggpubr)
library(ggrepel)
library(DT)
library(reshape2)
library(plotly)
library(tidyverse)
```


## Parameters

```{r}
# Paths
# path_to_mcl_cr_metric <- here::here("MCL/results/tables/cellranger_mapping/cellranger_mapping_metrics_mcl.csv")
path_to_mcl_cr_metric <- "~/Desktop/CNAG/mnt_clust/tonsil_atlas/current/MCL/results/tables/cellranger_mapping/cellranger_mapping_metrics_mcl.csv"
# path_to_sequencing_metadata <- here::here("MCL/1-cellranger_mapping/data/sequencing_metadata.csv")
path_to_sequencing_metadata <- "~/Desktop/CNAG/mnt_clust/tonsil_atlas/current/MCL/1-cellranger_mapping/data/sequencing_metadata.csv"

# Functions
# source(here::here("scRNA-seq/bin/utils.R"))
source("~/Desktop/CNAG/mnt_clust/tonsil_atlas/current/scRNA-seq/bin/utils.R")

barplot_data <- function(multiome_melted){
  multiome_melted$value <- round(multiome_melted$value, 2)
  ggbarplot(multiome_melted,
  x = "library_name",
  y = "value",
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
    facet_wrap(ncol = 1,scales = "free_y", ~variable) 
}
correlation_plot <- function(multiome_df, variable1, variable2) {
  ggscatter(multiome_df, x = variable1, y = variable2,
   add = "reg.line",  # Add regressin line
   add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
   conf.int = TRUE # Add confidence interval
   ) + stat_cor(method = "pearson")
}
```


## Load data


```{r}
# Load
cr_qc_metrics <- read_csv(path_to_mcl_cr_metric)
sequencing_metadata <- read_csv(path_to_sequencing_metadata, col_names = TRUE)


# Clean
colnames(cr_qc_metrics) <- str_replace_all(colnames(cr_qc_metrics), " ", "_")
sequencing_metadata_sub <- filter(sequencing_metadata, type == "RNA")
library_names <- sequencing_metadata_sub$library_name
library_names <- str_remove(library_names, "_GEX")
names(library_names) <- sequencing_metadata_sub$gem_id
cr_qc_metrics <- tibble::add_column(
  cr_qc_metrics,
  library_name = library_names,
  .before = "Sample_ID"
)
cr_qc_metrics$donor_id <- cr_qc_metrics$library_name %>%
  str_remove("(neg|pos)") %>%
  str_remove("_(1|2)")


# Print
print("QC summary table")
DT::datatable(cr_qc_metrics, options = list(scrollX = TRUE))


print("Sequencing metadata")
DT::datatable(sequencing_metadata, options = list(scrollX = TRUE))
```


# Overall statistics

## Estimated Number of Cells

Let us start by plotting the estimated number of cells per library. Note that this number will differ a lot from the final number of cells after applying future QC filters:

```{r}
num_cells_gg <- horizontal_barplot(
  cr_qc_metrics,
  categorical_var = "library_name",
  continuous_var = "Estimated_number_of_cells",
  ylab = "Estimated number of cells"
) +
  geom_col(aes(fill = donor_id))


print("Estimated number of cells:")
num_cells_gg
```


# Gene expression metrics

## Median genes per cell

```{r}
median_genes_gg <- horizontal_barplot(
  cr_qc_metrics,
  categorical_var = "library_name",
  continuous_var = "GEX_Median_genes_per_cell",
  ylab = "Median genes per cell"
) +
  geom_col(aes(fill = donor_id))


print("Median genes per cell:")
median_genes_gg
```


## Sequencing

Our first objective is to quantify the quality of our sequenced libraries prior to mapping. We will leverage the "Q30" variables in our dataset. According to 10X, these report the fraction of bases with a Q-score of at least 30 for different sequences: cell barcode, RNA reads, Unique Molecular Identifiers (UMI). Q-score is calculated as follows:

$$Q = -10\log10(p)$$
Where p is the probability of the base being wrongly called. Thus, bases with a high Q are highly reliable.

```{r fig.wide=TRUE}
# Sequencing QC
cols <- c("#cb3b25", "#b930df")


# Q30
q30_vars <- str_subset(colnames(cr_qc_metrics), "GEX_Q30")
q30_gg <- purrr::map(q30_vars, function(var) {
  p <- horizontal_barplot(
    cr_qc_metrics,
    categorical_var = "library_name",
    continuous_var = var,
    ylab = str_replace_all(var, "_",  " ")
  ) +
  geom_col(aes(fill = donor_id))
  p
})
q30_gg
```


# Mapping

Secondly, we will assess the quality of cellranger's mapping by comparing the percentage of reads mapping to the genome, intergenic regions, intronic and exonic regions across libraries. Reads mapping to intergenic regions suggest contamination of ambient DNA, while reads mapping to intronic regions may come from pre-mRNAs or mature splice isoforms that retain the introns:

```{r fig.wide=TRUE}
# Reads mapped to genome and transcriptome
mapping_qc_vars <- str_subset(colnames(cr_qc_metrics), "GEX_Reads_mapped")
mapping_qc_gg <- purrr::map(mapping_qc_vars, function(var) {
  p <- horizontal_barplot(
    cr_qc_metrics,
    categorical_var = "library_name",
    continuous_var = var,
    ylab = str_replace_all(var, "_",  " ")
  ) +
  geom_col(aes(fill = donor_id))
  p
})

mapping_qc_gg
```


# Chromatin accessibility metrics

## Basic stats

```{r}
num_peaks_gg <- horizontal_barplot(
  cr_qc_metrics,
  categorical_var = "library_name",
  continuous_var = "ATAC_Number_of_peaks",
  ylab = "ATAC Number of peaks"
) +
  geom_col(aes(fill = donor_id))
num_peaks_gg
```



## Sequencing

```{r}
selected_cols <- c(
  "library_name",
  "ATAC_Q30_bases_in_barcode",
  "ATAC_Q30_bases_in_read_1",
  "ATAC_Q30_bases_in_read_2"
)
sequencing_atac_melted <- melt(cr_qc_metrics[, selected_cols])
          
ggplot(sequencing_atac_melted, aes(variable, value)) +
  geom_boxplot(fill = NA, outlier.shape = NA) +
  geom_jitter(width = 0.1, height = 0) +
  geom_text_repel(data = sequencing_atac_melted, aes(label = library_name, fill = "black")) +
  theme_minimal() +
  theme(legend.position = "none",
      axis.title = element_text(size = 10),
      axis.text = element_text(size = 10, color = "black"))
```


## Valid barcodes
Fraction of read pairs with barcodes that match the whitelist after error correction. A value higher than 85% represent a high quality library.

```{r fig.wide = TRUE}
ggbarplot(cr_qc_metrics,
  x = "library_name",
  y =  "ATAC_Valid_barcodes",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
  geom_hline(yintercept  = 0.85, linetype='dotted', col = 'black')
```


## Transcription Start Site (TSS)
It is expected to obtain a large enrichment around TSS (> 5% EV, red dashed line), as these regions are known to show a high degree of chromatin accessibility compared to the flanking regions.

```{r fig.wide = TRUE}
data_tss <- cr_qc_metrics[, c("library_name", "ATAC_TSS_enrichment_score")]
data_tss.melt <- melt(data_tss)
barplot_data(data_tss.melt) +
  geom_hline(yintercept = 5, linetype = 2, color = "red")
```

# Joint Metrics
Here, we can visualize the number of the features (peaks and genes) and feature linkages detected. Note that the feature linkage is calculated taking in account genes with its proximal peaks and between pairs of proximal peaks across the genome (based on correlation). 


```{r fig.wide = TRUE}
barplot_data(melt(cr_qc_metrics[,c("library_name", "Linked_genes", "Linked_peaks")]))

linkage_melt <- melt(cr_qc_metrics[,c("library_name", "Feature_linkages_detected")])
barplot_data(multiome_melted = linkage_melt)
```


Correlation total number of genes and peaks

```{r}
correlation_plot(cr_qc_metrics, "ATAC_Number_of_peaks", "GEX_Total_genes_detected") +
 geom_text(aes(label = library_name))
```


# Session Info

```{r}
sessionInfo()
```
