---
title: 'Cellranger Mapping Multiome QC: additional cases Clinic'
author: "Ramon Massoni-Badosa"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction


# Introduction
The objective of this notebook is to perform a basic quality control of the mapping performed with [cellranger-arc](https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/what-is-cell-ranger-arc). Cell Ranger ARC analyzes data generated by the Chromium Single Cell Multiome ATAC + Gene Expression. Should note here, we will talk about features that can refer to a gene (declared from the hg38 reference transcriptome) or a peak (determined by the Peak Calling algorithm provided by Cell Ranger). 

The expected values (EV) included in the notebook are obtained from this [technical_note](https://assets.ctfassets.net/an68im79xiti/6uFMwMQwMySzluyIeBZhMl/e5acb853870c59d3f3fc826428cb6d3c/CG000368_TechnicalNote_InterpretingARCWebSummaryFilesforMultiomeATAC_GEX_RevA.pdf)

# Pre-processing

## Load packages

```{r}
library(tidyverse)
library(ggpubr)
library(plyr)
library(reshape2)
library(data.table)
library(knitr)
library(ggrepel)
library(here)
library(glue)
```


## Functions

```{r}
barplot_data <- function(multiome_melted){
  multiome_melted$value <- round(multiome_melted$value, 2)
  ggbarplot(multiome_melted,
  x = "library_name",
  y = "value",
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
    facet_wrap(ncol = 1,scales = "free_y", ~variable) 
}

correlation_plot <- function(multiome_df, variable1, variable2) {
  ggscatter(qc_samples, x = variable1, y = variable2,
     add = "reg.line",  # Add regressin line
     add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
     conf.int = TRUE # Add confidence interval
     ) + stat_cor(method = "pearson")
}
```


## Load data

```{r}
path_to_experiment3 <- here("multiome/1-cellranger_mapping/projects/experiment_3/jobs")
gem_ids <- list.dirs(path_to_experiment3, full.names = FALSE, recursive = FALSE)
qc_samples <- map_df(gem_ids, \(x) {
  path_to_file <- glue("{path_to_experiment3}/{x}/{x}/outs/summary.csv")
  df <- read_csv(path_to_file, col_names = TRUE)
  df
})
path_to_proj_metadata <- here("multiome/1-cellranger_mapping/data/tonsil_atlas_metadata_multiome.csv")
metadata <- read_csv(path_to_proj_metadata, col_names = TRUE)
metadata <- metadata[metadata$subproject %in% c("BCLLATLAS_133", "BCLLATLAS_134"), ]
DT::datatable(metadata)
```

```{r}
qc_samples$library_name <- revalue(
  qc_samples$`Sample ID`,
  c(
    "k8s4gkxa_phcaecsq"= "BCLL_25_T_1",
    "hagf3y16_sldip2ou" = "BCLL_25_T_2",
    "zoxefjul_5m23u91h" = "BCLL_26_T_GEX_1",
    "ng2o8d1y_g4kzaifb" = "BCLL_26_T_GEX_2"
  )
)
```


```{r}
DT::datatable(qc_samples, options = list(scrollX = TRUE))
```


#  Multiome Metrics Definitions 

## Estimated Number of Cells
Let us start by plotting the estimated number of cells per library. Note this  number will differ a lot from the final number of cells after applying future QC filters.

```{r fig.wide = TRUE}
barplot_data(melt(qc_samples[, c("library_name","Estimated number of cells")])) + coord_flip()
print(paste("Estimated number of cells for all the samples:", sum(qc_samples$`Estimated number of cells`)))
```


## Joint Metrics

Here, we can visualize the number of the features (peaks and genes) and feature linkages detected. Note that the feature linkage is calculated taking in account genes with its proximal peaks and between pairs of proximal peaks across the genome (based on correlation). 

```{r fig.wide = TRUE}
barplot_data(melt(qc_samples[,c("library_name",
                                "Linked genes",
                                "Linked peaks")]))

linkage_melt <- melt(qc_samples[,c("library_name",
                                   "Feature linkages detected")])
barplot_data(multiome_melted = linkage_melt)
```


# Chromatin Accessibility Metrics Definitions

## Sequencing 

Our first objective is to evaluated the quality and the quantity of our sequenced libraries prior to mapping.

### Sequenced read pairs

Here, you can see the total number of sequenced read pairs assigned to the Multiome ATAC library. The optimal number of read pairs are 25,000 per cell taking in account the technical note.

```{r fig.wide = TRUE}
qc_samples$`ATAC Mean raw read pairs per cell` <- round(qc_samples$`ATAC Mean raw read pairs per cell`, 2)
ggbarplot(qc_samples,
  x = "library_name",
  y = "ATAC Mean raw read pairs per cell",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
  geom_hline(yintercept  = 25000, linetype='dotted', col = 'black')
```

Here, you can see the quality (leveraging the "Q30" variables)  of sequenced read pairs assigned to the Multiome ATAC library.
 
```{r fig.wide = TRUE}
multiome_melted <- melt(qc_samples[,c("library_name", "ATAC Q30 bases in barcode",
"ATAC Q30 bases in read 1",        
"ATAC Q30 bases in read 2")])
          
ggplot(multiome_melted, aes(variable, value)) +
  geom_jitter(width = 0.1, height = 0) +
  geom_text_repel(data = multiome_melted, aes(label = library_name, fill = "black")) +
  theme_minimal() +
  theme(legend.position = "none",
      axis.title = element_text(size = 10),
      axis.text = element_text(size = 10, color = "black"))
```


### Valid barcodes

Fraction of read pairs with barcodes that match the whitelist after error correction. A value higher than 85% represent a high quality library.

```{r fig.wide = TRUE}
ggbarplot(qc_samples,
  x = "library_name",
  y =  "ATAC Valid barcodes",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
  geom_hline(yintercept  = 0.85, linetype='dotted', col = 'black')
```


### Percent of duplicates

The percentatge of the duplicated data is correlated with the library size and the sequencing depth.

```{r fig.wide = TRUE}
correlation_plot(qc_samples, "ATAC Percent duplicates", "ATAC Sequenced read pairs")
correlation_plot(qc_samples, "ATAC Percent duplicates", "ATAC Number of peaks")
```


## Cells

###  Median number of fragments per cell
High-Quality Fragment: A read-pair with mapping quality > 30, that is not chimerically mapped, has a valid 10x barcode, and maps to any nuclear contig (not mitochondria) that contains at least one gene.

```{r  fig.wide = TRUE}
qc_atac_read_pairs <- melt(qc_samples[,c("library_name","ATAC Mean raw read pairs per cell",                      "ATAC Median high-quality fragments per cell")])
barplot_data(qc_atac_read_pairs)
qc_atac_fraction <- melt(qc_samples[,c("library_name","ATAC Fraction of high-quality fragments in cells","ATAC Fraction of transposition events in peaks in cells")])
barplot_data(qc_atac_fraction)
```
          

### Targeting Metrics

The targetting metrics can be summarized by these 4 main score:

#### Total number of peaks on primary contigs. 

This number presents a high correlation with the sequencing depth specifically with the high quality fragments.
 
```{r fig.wide = TRUE}
ggbarplot(qc_samples, "library_name", "ATAC Number of peaks",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none"),
  position = position_dodge(0.9))
```


#### Fraction of high quality fragments that overlapp in peaks.

The fraction of high quality fragments in cell are expected to be higher 40%. The fraction of transposition events that fall within peaks > 25%

```{r fig.wide = TRUE}
multiome_melted <- melt(qc_samples[,c("library_name", "ATAC Fraction of high-quality fragments in cells",
"ATAC Fraction of high-quality fragments overlapping peaks", "ATAC Fraction of high-quality fragments overlapping TSS")])
                 
multiome_melted$value <- round(multiome_melted$value,2)                   
ggbarplot(multiome_melted, "library_name", "value",
  fill = "variable", color = "variable", palette = "Paired",
  label = TRUE,
  ggtheme = theme_pubr(x.text.angle = 45),
  position = position_dodge(0.9))
```


#### Fraction of genome in peaks

```{r fig.wide = TRUE}
ggplot(melt(qc_samples[,c("library_name",
                          "ATAC Fraction of genome in peaks")]), aes(variable, value)) +
  geom_jitter(width = 0.1, height = 0) +
  theme_minimal() +
  theme(legend.position = "none",
      axis.title = element_text(size = 10),
      axis.text = element_text(size = 10, color = "black"))
```

#### Transcription Start Site (TSS)

It is expected to obtain a large enrichment around TSS (> 5% EV, red dashed line), as these regions are known to show a high degree of chromatin accessibility compared to the flanking regions.

```{r fig.wide = TRUE}
data_tss <- qc_samples[, c("library_name", "ATAC TSS enrichment score")]
data_tss.melt <- melt(data_tss)
barplot_data(data_tss.melt) + geom_hline(yintercept = 5, linetype = 2, color = "red")
```


# Gene Expression Metrics Definitions

## Median genes per cell

```{r  fig.wide = TRUE}
median_genes_gg <- melt(qc_samples[, c("library_name","GEX Median genes per cell")])
barplot_data(median_genes_gg)
```


### Sequenced read pairs
Here, you can see the total number of sequenced read pairs assigned to the Multiome Gene Expression library. The optimal number of read pairs are ~20,000 per cell taking in account the technical note.

```{r fig.wide = TRUE}
qc_samples$`GEX Mean raw reads per cell` <- round(qc_samples$`GEX Mean raw reads per cell`, 2)
ggbarplot(qc_samples,
  x = "library_name",
  y =  "GEX Mean raw reads per cell",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45)) + 
  geom_hline(yintercept  = 20000, linetype='dotted', col = 'black')
```

Here, you can see the quality (leveraging the "Q30" variables)  of sequenced read pairs assigned to the Multiome Gene Expresion library.
 
```{r fig.wide = TRUE}
multiome_melted <- melt(qc_samples[,c("library_name", "GEX Q30 bases in barcode",
"GEX Q30 bases in UMI",        
"GEX Q30 bases in read 2")])


ggplot(multiome_melted, aes(variable, value)) +
  geom_jitter(width = 0.1, height = 0) +
  geom_text_repel(data = multiome_melted, aes(label = library_name, fill = "black")) +
  theme_minimal() +
  theme(legend.position = "none",
      axis.title = element_text(size = 10),
      axis.text = element_text(size = 10, color = "black"))
```


### Valid barcodes

Fraction of read pairs with barcodes that match the whitelist after error correction. A value higher than 80% represent a high quality library,

```{r fig.wide = TRUE}
ggbarplot(qc_samples,
  x = "library_name",
  y =  "GEX Valid barcodes",
  label = TRUE,
  fill = "gray70",
  ggtheme = theme_pubr(x.text.angle = 45,legend = "none")) + 
  geom_hline(yintercept  = 0.85, linetype='dotted', col = 'black')
```

### Percent of duplicates

The percentatge of the duplicated data is correlated with the library size and the sequencing depth.
```{r fig.wide = TRUE}
correlation_plot(qc_samples,"GEX Percent duplicates","GEX Sequenced read pairs")
correlation_plot(qc_samples,"GEX Percent duplicates","GEX Total genes detected")
```

## Mapping
Secondly, we will assess the quality of cellranger's mapping by comparing the percentage of reads mapping to the genome, intergenic regions, intronic and exonic regions across libraries. Reads mapping to intergenic regions suggest contamination of ambient DNA, while reads mapping to intronic regions may come from pre-mRNAs or mature splice isoforms that retain the introns.

The fraction of sequenced reads that map to a unique gene in the transcriptome is expected to be higher than the 50%.

```{r fig.wide = TRUE}
qc_samples$`GEX Fraction of transcriptomic reads in cells` <- round(qc_samples$`GEX Fraction of transcriptomic reads in cells`, 2)
ggbarplot(qc_samples,
  x = "library_name",
  y =  "GEX Fraction of transcriptomic reads in cells",
  fill = "gray70",
  label = TRUE,
  ggtheme = theme_pubr(x.text.angle = 45)) + 
  geom_hline(yintercept  = 0.50, linetype='dotted', col = 'black')
```

```{r fig.wide=TRUE}
multiome_melted <- melt(qc_samples[,c("library_name", "GEX Reads mapped to genome",  "GEX Reads mapped confidently to genome")])
multiome_melted$value <- round(multiome_melted$value,2)                    
ggbarplot(multiome_melted, 
          x = "library_name", 
          y = "value",
          fill = "variable", palette = "Paired",  label = TRUE,
  ggtheme = theme_pubr(x.text.angle = 45),
  position = position_dodge(0.9))
```

The fraction of sequenced reads that map to a unique gene in the transcriptome is expected to be higher than 50% (blue dashed line). However, the percentatge of reads mapped confidently to intergenic regions and reads mapped antisense to gene are expected to be lower than 30% (red dashed line).

```{r fig.wide=TRUE, fig.height=12}
mapping_qc_vars <- c(
  "library_name",
  "GEX Reads mapped confidently to intergenic regions",
  "GEX Reads mapped confidently to intronic regions",
  "GEX Reads mapped confidently to exonic regions",
  "GEX Reads mapped confidently to transcriptome",
  "GEX Reads mapped antisense to gene")

mapping_qc_gg.melt <- melt(qc_samples[,mapping_qc_vars])

ggplot(mapping_qc_gg.melt, aes(variable, value)) +
  geom_jitter(width = 0.1, height = 0) +
  theme_minimal() +
  theme(legend.position = "none",
      axis.title = element_text(size = 12),
      axis.text = element_text(size = 12, color = "black")) + coord_flip() + 
  geom_hline(yintercept = c(0.3,0.5), col = c("red","blue"),  linetype='dotted')
```


# Sequencing Saturation

Thirdly, we will plot the number of detected genes per library as a function of the total reads sequenced. We know that this function reaches a plateau, whereby more sequenced reads does not result in more detected genes. In those scenarios, we say that we have sequenced until saturation:

```{r fig.width = 8, fig.height = 6}
ggplot(qc_samples, aes(`GEX Sequenced read pairs`, `GEX Total genes detected`)) +
    geom_point() +
    geom_text_repel(data = qc_samples, aes(label = library_name), color = "black") +
    labs(x = "Number of Reads", y = "Total Genes Detected", color = "") +theme_classic() +
    theme(axis.title = element_text(size = 13, color = "black"),
          axis.text = element_text(size = 12, color = "black"),
          legend.text = element_text(size = 12, color = "black"))
```

```{r}
sessionInfo()
```
