---
title: "Quality control and filtering"
author: "Paula Soler-Vila"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```

# Objective

The objective of this notebook is to create a Seurat object, assess the quality of the data and filter out the low-quality cells.

# Pre-processing

## Load packages

```{r}
library(Signac)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(tidyverse)
library(plyr)
library(reshape2)
library(data.table)
library(GenomicRanges)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(hdf5r)
library(stringr)
set.seed(173)
```

## Parameters
```{r}
# Paths
path_to_data <- here::here("scATAC-seq/results/Aggregated/")
path_to_save <- here::here("scATAC-seq/results/R_objects/")
path_to_scrublet <- here::here("scATAC-seq/results/tables/scrublet/")
path_to_donor_metadata <- here::here("data/tonsil_atlas_donor_metadata.csv")
path_to_proj_metadata <- here::here("scATAC-seq/1-cellranger_mapping/data/tonsil_atlas_metadata_atac.csv")

# Thresholds
min_peak_region_fragments <- 700 
TSS_enrichment <- 2 
blacklist_rat <- 0.03 
max_peak_region_fragments <- 30000 
pct_in_peaks <- 15
min_cells <- 5
min_lib_size_atac <- 500
max_lib_size_atac <- 100000
```

## Gene annotation
Extraction of gene annotations from EnsDb using hg38 as the reference assembly.

```{r comment=FALSE}
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- "UCSC"
genome(annotations) <- "hg38"
```

# Load data
Signac uses information from two related input files (created using CellRanger):

1. Peak/Cell matrix
2. Fragment file

A Seurat object is created using the filtered peak/cell matrix (in h5 format) and cell metadata generated by cellranger-atac (singlecell.csv file), and store the path to the fragment file on disk in the Seurat object. 

```{r}
read_atac_data <- function(url_project) {
  counts <- Seurat::Read10X_h5(filename = paste0(url_project, "filtered_peak_bc_matrix.h5"))

  metadata <- read.csv(
    file = paste0(url_project, "singlecell.csv"),
    header = TRUE, row.names = 1
  )

  chrom_assay <- Signac::CreateChromatinAssay(
    counts = counts,
    sep = c(":", "-"),
    genome = "hg38",
    fragments = paste0(url_project, "fragments.tsv.gz"),
    min.cells = min_cells
  )

  Tonsil <- Seurat::CreateSeuratObject(
    counts = chrom_assay,
    assay = "ATAC",
    meta.data = metadata
  )

  #  Adding gene annotation to the Seurat object fot the hg38 human genome.
  Annotation(Tonsil) <- annotations
  return(Tonsil)
}
```

```{r}
# Read samples and assign them to a proper gem_id name
tonsil_aggregated <- read_atac_data(path_to_data)
tonsil_aggregated@meta.data$lib_number <- as.data.frame(str_split(row.names(tonsil_aggregated@meta.data), "-", simplify = TRUE))$V2

tonsil_aggregated@meta.data$gem_id <- revalue(
  tonsil_aggregated@meta.data$lib_number,
  c(
    "1" = "jbp85ve3_3179tpob",
    "2" = "sy59df1i_56bwpc10",
    "3" = "aks0q3tw_lsr4h73j",
    "4" = "fnpetv1y_agr9h7v4",
    "5" = "jmjpvq6z_fr6c4nyp",
    "6" = "yja17lfh_kdslhq0g",
    "7" = "u51p7uvm_s5fhku31",
    "8" = "vcan8qd9_sr5u40yg",
    "9" = "dhqzxpdk_h8em75r9",
    "10" = "lacrul5e_67e3wsyf",
    "11" = "lzj8c3xo_ytz0xw8m",
    "12" = "m7maqeew_gt2mrtf3",
    "13" = "xazsjcvt_kzi9e2rf"
  )
)
```

```{r}
# Read metadata and add it taking in account the gem_id 
donor_metadata <- read_csv(path_to_donor_metadata, col_names = TRUE)
atac_metadata <- read_csv(path_to_proj_metadata, col_names = TRUE)
atac_metadata$technique <- c("NON_FACS","NON_FACS","FACS",
                             "FACS","NON_FACS","NON_FACS",
                             "FACS","FACS","NON_FACS",
                             "NON_FACS","NON_FACS","FACS","FACS")

donor_ids <- atac_metadata$donor_id
names(donor_ids) <- atac_metadata$gem_id
tonsil_aggregated@meta.data$donor_id <- donor_ids[tonsil_aggregated$gem_id]

library_ids <- atac_metadata$library_name
names(library_ids) <- atac_metadata$gem_id
tonsil_aggregated@meta.data$library_name <- library_ids[tonsil_aggregated$gem_id]

technique_ids <- atac_metadata$technique
names(technique_ids) <- atac_metadata$gem_id
tonsil_aggregated@meta.data$technique <- technique_ids[tonsil_aggregated$gem_id]

sex_vec <- donor_metadata$sex
age_vec <- donor_metadata$age
age_group_vec <- donor_metadata$age_group
hospital_vec <- donor_metadata$hospital
names(sex_vec) <- donor_metadata$donor_id
names(age_vec) <- donor_metadata$donor_id
names(age_group_vec) <- donor_metadata$donor_id
names(hospital_vec) <- donor_metadata$donor_id
tonsil_aggregated@meta.data$sex <- sex_vec[tonsil_aggregated$donor_id]
tonsil_aggregated@meta.data$age <- age_vec[tonsil_aggregated$donor_id]
tonsil_aggregated@meta.data$age_group <- age_group_vec[tonsil_aggregated$donor_id]
tonsil_aggregated@meta.data$hospital <- hospital_vec[tonsil_aggregated$donor_id]
tonsil_aggregated@meta.data$assay <- "atac"

tonsil_aggregated[["ATAC"]]
```

```{r}
metadata_libraries <- tonsil_aggregated@meta.data
```

## General biological features of the experiments

```{r fig.wide=TRUE}

na <- ggbarplot(melt(table(metadata_libraries$age_group)), "Var1", "value",
  lab.size = 5, color = "Var1", fill = "Var1",
  label = TRUE, lab.pos = "in", title = "Number of cells per age_group diversity"
)
na
```

```{r fig.wide=TRUE}

ng <- ggbarplot(melt(table(metadata_libraries$sex)), "Var1", "value",
  lab.size = 5, color = "Var1", fill = "Var1",
  label = TRUE, lab.pos = "in", title = "Number of cells per sex diversity"
)
ng
```

```{r fig.wide=TRUE}

nt <- ggbarplot(melt(table(metadata_libraries$technique)), "Var1", "value",
  x.text.angle = 45,
  lab.size = 5, color = "Var1", fill = "Var1",
  label = TRUE, lab.pos = "in", title = "Number of cells per tecnhique"
)
nt
```

## General statistics of the experiments
We compute some general parameters to have a broad overview of the libraries' features that we have. Specifically, we check the number of the cells, the median of the fragments (and fragments that overlap in peaks) per library_name and the library_name complexity (ncount ATAC).

```{r fig.wide=TRUE}

nc <- ggbarplot(melt(table(metadata_libraries$library_name)), "Var1", "value",
  x.text.angle = 45,
  lab.size = 5, fill = "gray70",
  label = TRUE, lab.pos = "in", title = "Number of cells"
)
nc
```

```{r fig.wide=TRUE}
mr <- ggbarplot(melt(tapply(metadata_libraries$passed_filters, metadata_libraries$library_name, median)),
  "Var1", "value",
  x.text.angle = 45,
  lab.size = 5, fill = "gray70", 
  label = TRUE, lab.pos = "in", title = "Median of read-pairs(fragments)"
)
mr
```

```{r fig.wide=TRUE}
mf <- ggbarplot(melt(tapply(metadata_libraries$peak_region_fragments, metadata_libraries$library_name, median)),
  "Var1", "value",
  x.text.angle = 45,
  lab.size = 5, fill = "gray70", 
  label = TRUE, lab.pos = "in", title = "Median of number of fragments overlapping peaks"
)
mf
```

# Quality control

We compute and save four QC metrics to assess data quality.

```{r}
quality_control <- function(seurat_object) {
  seurat_object <- NucleosomeSignal(object = seurat_object)

  seurat_object$pct_reads_in_peaks <- seurat_object$peak_region_fragments / seurat_object$passed_filters * 100

  seurat_object$blacklist_ratio <- seurat_object$blacklist_region_fragments / seurat_object$peak_region_fragments

  seurat_object <- Signac::TSSEnrichment(object = seurat_object, fast = FALSE)

  seurat_object$high.tss <- ifelse(seurat_object$TSS.enrichment > 2, "High", "Low")

  return(seurat_object)
}
```

```{r}
tonsil_aggregated_QC <- quality_control(tonsil_aggregated)
saveRDS(tonsil_aggregated_QC, paste0(path_to_save,"1.tonsil_aggregated_QC.rds"))
metadata_libraries <- tonsil_aggregated_QC@meta.data
```

## Nucleosome banding pattern
Quantification of the ratio of mononucleosomal (fragment lengths between 147 and 294bp) to nucleosome-free fragments(less than 147)

```{r}
nbp <- function(seurat_object) {
  f <- FragmentHistogram(object = seurat_object)
  print(head(f$data))

  NFR <- length(which(f$data$length < 147)) / nrow(f$data) * 100
  MONO <- length(which(f$data$length > 147 & f$data$length < 294)) / nrow(f$data) * 100
  DI <- length(which(f$data$length > 294)) / nrow(f$data) * 100

  min.threshold <- 147
  max.threshold <- 294

  # options(repr.plot.width=7, repr.plot.height=2)
  options(repr.plot.width = 17, repr.plot.height = 8)
  p <- ggplot(f$data, aes(length)) + # ggtitle(unique(seurat_object$library_name)) +
    geom_histogram(binwidth = 1, alpha = .5, fill = "blue") +
    geom_density(aes(y = ..count..), bw = 1, alpha = 0, col = "black", lwd = 1) +
    scale_x_continuous(limits = c(0, 500)) +
    geom_vline(xintercept = c(min.threshold, max.threshold)) +
    theme_minimal() +
    geom_vline(xintercept = 39, color = "red") +
    geom_text(x = 80, y = 20, label = round(NFR, 2), size = 8) +
    geom_text(x = 200, y = 20, label = round(MONO, 2), size = 8) +
    geom_text(x = 350, y = 20, label = round(DI, 2), size = 8)
  print(p)
}
```

```{r fig.wide=TRUE}
plot_nbp <- nbp(tonsil_aggregated_QC)
```

```{r fig.wide=TRUE}
nsig <- ggviolin(metadata_libraries,
  x = "library_name", x.text.angle = 45,
  y = "nucleosome_signal", fill = "gray70") + scale_y_log10() + scale_y_log10()
nsig
```

## Transcriptional start site (TSS) enrichment score

```{r fig.wide=TRUE}
tss_signac <- Signac::TSSPlot(tonsil_aggregated_QC, group.by = "high.tss") + ggtitle("TSS enrichment score") + NoLegend()

tss <- ggviolin(metadata_libraries,
  x = "library_name", x.text.angle = 45,
  y = "TSS.enrichment", fill = "gray70") + 
  geom_hline(yintercept = TSS_enrichment, linetype='dashed', col = 'black') + scale_y_log10()
tss
```

## Fractions of fragments that fall within ATAC-seq peaks

```{r fig.wide=TRUE}
pct <- ggviolin(metadata_libraries,
  x = "library_name", fill = "gray70",
  y = "pct_reads_in_peaks", x.text.angle = 45
) + geom_hline(yintercept = pct_in_peaks, linetype='dashed', col = 'black')
pct
```

## Total number of fragments in peaks

```{r fig.wide=TRUE}
peak <- ggviolin(metadata_libraries,
  x = "library_name", fill = "gray70",
  y = "peak_region_fragments", x.text.angle = 45
) + scale_y_log10() +
  geom_hline(yintercept = c(min_peak_region_fragments, max_peak_region_fragments), linetype='dashed', col = 'black')
peak
```

## Ratio reads in ‘blacklist’ sites

```{r fig.wide=TRUE}
black <- ggviolin(metadata_libraries,
  x = "library_name",  x.text.angle = 45, fill = "gray70",
  y = "blacklist_ratio") + scale_y_log10() +
  geom_hline(yintercept = blacklist_rat, linetype='dashed', col = 'black')
black
```

## Library size

```{r fig.wide=TRUE}
ls <- ggviolin(metadata_libraries,
  x = "library_name", x.text.angle = 45, fill = "gray70",
  y = "nCount_ATAC") +  scale_y_log10() + 
  geom_hline(yintercept = c(min_lib_size_atac,max_lib_size_atac), linetype='dashed', col = 'black')
ls
```

# Read scrublet predictions
```{r}
files_scrublet <- list.files(path_to_scrublet)
scrublet_gem_ids <- files_scrublet %>%
  str_remove("scrublet_doublet_prediction-") %>%
  str_remove(".csv")

df <- list.files(path_to_scrublet, full.names = T) %>%
  setNames(nm = scrublet_gem_ids) %>%
  map_df(~ read_csv(.x, col_types = cols(), skip = 1, col_names = FALSE), .id = "file_name")

colnames(df) <- c("gem_id", "barcode", "scrublet_doublet_scores", "scrublet_predicted_doublet")

tonsil_aggregated_QC@meta.data$barcode_aggreg <- row.names(tonsil_aggregated_QC@meta.data)
tonsil_aggregated_QC@meta.data$barcode <- gsub('[0-9]+',"1",row.names(tonsil_aggregated_QC@meta.data))
tonsil_aggregated_QC@meta.data <- merge(df,tonsil_aggregated_QC@meta.data, by = c("gem_id","barcode"))
row.names(tonsil_aggregated_QC@meta.data) <- tonsil_aggregated_QC@meta.data$barcode_aggreg

tonsil_aggregated_QC@meta.data$barcode <- NULL
tonsil_aggregated_QC@meta.data$barcode_aggreg <- NULL
```

# Filtering
Finally we remove cells that are outliers for these QC metrics. Also, we decided to remove the cells from young and old adults for the sample size.
```{r}
filtering_QC <- function(seurat_object) {
  seurat_object <- subset(
    x = seurat_object, peak_region_fragments > min_peak_region_fragments &
      TSS.enrichment > TSS_enrichment &
      blacklist_ratio < blacklist_rat &
      peak_region_fragments < max_peak_region_fragments &
      pct_reads_in_peaks > pct_in_peaks &
      nCount_ATAC < max_lib_size_atac &
      nCount_ATAC > min_lib_size_atac
  )

  return(seurat_object)
}
```

```{r}
tonsil_aggregated_filtered <- filtering_QC(tonsil_aggregated_QC)
saveRDS(tonsil_aggregated_filtered, paste0(path_to_save, "2.tonsil_aggregated_filtered.rds"))
```

```{r fig.wide=TRUE}
raw <- melt(table(tonsil_aggregated_QC$library_name))
filt <- melt(table(tonsil_aggregated_filtered@meta.data$library_name))
comparison <- merge(raw,filt, by = "Var1")
colnames(comparison) <- c("library_name","Raw","Filt")

ggbarplot(melt(comparison), "library_name", "value", fill = "variable",
  x.text.angle = 45, lab.size = 5, color = "variable", position = position_dodge(0.9),
  label = TRUE, lab.pos = "in", title = "Number of  cells")
```

```{r fig.wide=TRUE}
number_of_cells <- data.frame(
  c("Aggregated_raw", "Aggregated_filtered"),
  c(sum(raw$value), sum(filt$value)))
colnames(number_of_cells) <- c("samples", "n_cells")
ggbarplot(number_of_cells, "samples", "n_cells",
  x.text.angle = 45,
  lab.size = 5, color = "samples", fill = "samples",
  label = TRUE, lab.pos = "in"
)
```

# Session Info
```{r}
sessionInfo()
```
