---
title: "Filter out poor-quality cells (scRNA-seq)"
author: "Ramon Massoni-Badosa"
date: "`r format(Sys.time(), '%B %Y')`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction

Here we will initialize the Seurat objects for the scRNA-seq slots and filter out poor quality cells. Since we have to execute the code locally (HPC is down), we will split the analysis into many steps so R can tolerate it. This will include

1. Merge the count matrices for donors 102 and 403 independently. We will do that for scRNA-seq and scATAC-seq separately and save the corresponding Seurat objects.
2. Calculate the QC metrics for each of them and filter out poor-quality cells.
3. Merge both assays into a single one and perform downstream analysis.


# Pre-processing

## Load packages

```{r}
library(Seurat)
library(ggpubr)
library(tidyverse)
set.seed(173)
```


## Define paths and parameters

```{r}
# Paths
path_to_data <- here::here("MCL/1-cellranger_mapping/projects/experiment_1/jobs")
paths_to_files <- list.files(
  path_to_data,
  pattern = "filtered_feature_bc_matrix.h5",
  recursive = TRUE,
  full.names = TRUE
)
gem_ids <- list.dirs(path_to_data, full.names = FALSE, recursive = FALSE)
names(paths_to_files) <- gem_ids
path_to_sequencing_metadata <- here::here("MCL/1-cellranger_mapping/data/sequencing_metadata.csv")
path_to_save_102 <- here::here("MCL/results/R_objects/1.seurat_object_filtered_102_rna.rds")
path_to_save_413 <- here::here("MCL/results/R_objects/1.seurat_object_filtered_413_rna.rds")


# Thresholds
min_n_counts_rna <- 900
max_n_counts_rna <- 25000
max_pct <- 10
```


# Load data and create Seurat object

```{r}
# Read data
counts <- purrr::map(paths_to_files, Seurat::Read10X_h5)


# Create a Seurat object containing the scRNA data
seurat_list <- purrr::map2(counts, gem_ids, function(l, x) {
  seurat_obj <- Seurat::CreateSeuratObject(
    counts = l$`Gene Expression`,
    assay = "RNA"
  )
  seurat_obj$gem_id <- x
  seurat_obj
})
names(seurat_list) <- names(paths_to_files)


# Read sequencing metadata
sequencing_metadata <- read_csv(path_to_sequencing_metadata)
DT::datatable(sequencing_metadata, options = list(scrollX = TRUE))
```


# Merge objects

```{r}
seurat_102 <- merge(
  x = seurat_list$kpi1zxjo_542l3n1f,
  y = c(seurat_list$poba5cxi_qte0l4uh, seurat_list$rvnas5hf_wfkqdmhy)
)
seurat_413 <- merge(
  x = seurat_list$jt18zlyw_fy97xnbm,
  y = c(seurat_list$wsz3j4g1_27jak0zo, seurat_list$ywt2m1v6_16u4fqig)
)
rm(seurat_list)
```


# Define metadata

```{r}
sequencing_metadata <- sequencing_metadata[sequencing_metadata$type == "RNA", ]


# 102
new_metadata_102 <- left_join(
  seurat_102@meta.data,
  sequencing_metadata,
  by = "gem_id"
)
rownames(new_metadata_102) <- colnames(seurat_102)
seurat_102@meta.data <- new_metadata_102


# 413
new_metadata_413 <- left_join(
  seurat_413@meta.data,
  sequencing_metadata,
  by = "gem_id"
)
rownames(new_metadata_413) <- colnames(seurat_413)
seurat_413@meta.data <- new_metadata_413
```


# Calculate basic QC metrics

```{r}
# Percentage of mitochondrial expression
seurat_413$pct_mt <- PercentageFeatureSet(
  seurat_413,
  pattern = "^MT-"
)
seurat_102$pct_mt <- PercentageFeatureSet(
  seurat_102,
  pattern = "^MT-"
)


# Percentage of ribosomal expression
seurat_413$pct_mt <- PercentageFeatureSet(
  seurat_413,
  pattern = "^RPS"
)
seurat_102$pct_mt <- PercentageFeatureSet(
  seurat_102,
  pattern = "^RPS"
)
```


# Filter poor-quality cells

## Visualize

```{r}
# RNA
qc_df <- bind_rows(seurat_102@meta.data, seurat_413@meta.data)
qc_metrics_rna <- c("nCount_RNA", "nFeature_RNA", "pct_mt")
qc_ggs_rna <- purrr::map(qc_metrics_rna, function(x) {
  p <- ggplot(qc_df, aes_string("library_name", x, fill = "donor_id")) +
    geom_violin() +
    xlab("") +
    theme_classic() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1))
  p
})
qc_ggs_rna[[1]] <- qc_ggs_rna[[1]] +
  scale_y_log10() +
  geom_hline(yintercept = min_n_counts_rna, color = "red", linetype = "dashed") +
  geom_hline(yintercept = max_n_counts_rna, color = "red", linetype = "dashed")
qc_ggs_rna[[3]] <- qc_ggs_rna[[3]] +
  geom_hline(yintercept = max_pct, color = "red", linetype = "dashed")
ggarrange(plotlist = qc_ggs_rna, ncol = 3, common.legend = TRUE)
```

## Subset

```{r}
# 102
seurat_102
seurat_102 <- subset(
x = seurat_102,
  subset = nCount_RNA < max_n_counts_rna &
    pct_mt < max_pct &
    nCount_RNA > min_n_counts_rna
)
seurat_102


# 413
seurat_413
seurat_413 <- subset(
x = seurat_413,
  subset = nCount_RNA < max_n_counts_rna &
    pct_mt < max_pct &
    nCount_RNA > min_n_counts_rna
)
seurat_413
```

# Save

```{r}
saveRDS(seurat_102, path_to_save_102)
saveRDS(seurat_413, path_to_save_413)
```

