---
author: "Marc Elosua-Bayes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
title: "Epithelial cells SC mapping"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE, cache = FALSE,
                      cache.lazy = FALSE)
options(width = 1200)
set.seed(123)
```

## Introduction
In this script we are going to map myeloid cell subtypes onto the Visium slides.

## Libraries
```{r}
library(Seurat)
library(ggpubr)
library(cowplot)
library(dplyr)
library(ggplot2)
library(stringr)
library(readr)
library(SPOTlight)
```

## Setting parameters
Loading necessary paths and parameters
```{r}
set.seed(123)
source(here::here("misc/paths.R"))
source(here::here("utils/bin.R"))

"{epithelium}/{plt_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = ,
             showWarnings = FALSE,
             recursive = TRUE)

"{epithelium}/{robj_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = ,
             showWarnings = FALSE,
             recursive = TRUE)
```

Extract sample id and get Donor ID
```{r}
# sample_id <- params$sample_id
sample_id <- "esvq52_nluss5"
donor_id <- id_sp_df[id_sp_df$gem_id == sample_id, ] %>% dplyr::pull(donor_id)
```

## Load data
We have 4 different datasets that we are going to analyze separately.
The spatial data comes from the script *03-clustering/03-clustering_integration.Rmd* while the sc data can be found in Ramon's scRNAseq analysis: **/scratch/devel/rmassoni/tonsil_atlas_private/2-DOWNSTREAM_PROCESSING/results/R_objects/processed_seurat_objects/processed_seurat_objects/tonsil_integrated_with_harmony_scrublet_annotated.rds**. \
```{r}
sp_obj <- "{clust}/{robj_dir}/integrated_spatial.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)

# Single cell data
sc_obj <- "{decon}/{robj_dir}/tonsil_integrated_with_harmony_scrublet_annotated.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)

# level 5 Epithelium T Cells
lvl4_epi_obj <- "{epithelium}/{robj_dir}/epithelial_annotated_level_4.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

## Analysis

### Metadata
We are going to use SPOTlight to deconvolute the spatial spots and map the cell types.

We will start by adding the level5 definitive metadata labels to the full object.
```{r}
lvl4_metadata <- lvl4_epi_obj@meta.data %>%
  dplyr::select(barcode, annotation_level_3)

table(lvl4_metadata$annotation_level_3)

sc_obj@meta.data <- sc_obj@meta.data %>%
  tibble::rownames_to_column("barcode") %>%
  dplyr::left_join(lvl4_metadata, by = "barcode") %>%
  dplyr::mutate(
    # Annotation for SPOTlight
    annotation_deconv = dplyr::if_else(
      is.na(annotation_level_3), annotation_level1, annotation_level_3)
    # Annotation for SPOTlight with condensed names
    # annotation_cons_deconv = dplyr::case_when(
    #   annotation_deconv %in% c("Central Mem PASK-", "Central Mem PASK+") ~ "Central Mem",
    #   annotation_deconv %in% c("Follicular Th CXCL13+CBLB+_0", "Follicular Th CXCL13+CBLB+_1", "Follicular Th CXCL13+CBLB+_2") ~ "Follicular Th CXCL13+",
    #   annotation_deconv %in% c("Central Mem PASK-", "Central Mem PASK+") ~ "Central Mem PASK",
    #   TRUE ~ annotation_deconv)
    ) %>%
  # Remove CD4 T cells not anotated in Level 5 since they have been removed in the process
  dplyr::filter(! annotation_deconv %in% c("epithelial", "unknown")) %>%
  tibble::column_to_rownames("barcode")
```

### Marker calculation
Since the normalization was carried out with the entire dataset we are going to do normalization again within this subset
```{r}
sc_obj <- Seurat::NormalizeData(
  sc_obj,
  normalization.method = "LogNormalize",
  scale.factor = 10000,
  verbose = FALSE) %>%
  Seurat::FindVariableFeatures(., nfeatures = 3000, verbose = FALSE) %>% 
  Seurat::ScaleData(., verbose = FALSE)
```

1st we need to determine the marker genes for each high-level cluster
```{r}
markers_lvl1_out <- "{epithelium}/{robj_dir}/markers_lvl1_epith.RDS" %>% 
  glue::glue() %>% 
  here::here()

if (file.exists(markers_lvl1_out)) {
  
  markers_lvl1 <- readRDS(file = markers_lvl1_out)
  
} else {
  #### Extract marker genes from each cluster ####
  Seurat::Idents(object = sc_obj) <- sc_obj@meta.data[, "annotation_level1"]
  markers_lvl1 <- Seurat::FindAllMarkers(
    object = sc_obj, 
    assay = "RNA",
    slot = "data",
    verbose = TRUE, 
    only.pos = TRUE,
    max.cells.per.ident = 200)
  
  saveRDS(object = markers_lvl1, file = markers_lvl1_out)
}
DT::datatable(markers_lvl1, filter = "top")
```

Now find the markers for CD4 T cells level 5 only between CD4 T cells
```{r}
markers_lvl4_out <- "{epithelium}/{robj_dir}/markers_lvl4_epithelial.RDS" %>% 
  glue::glue() %>% 
  here::here()

if (file.exists(markers_lvl4_out)) {
  
  markers_lvl4 <- readRDS(file = markers_lvl4_out)
  
} else {
  # lvl4 epithelium cells lvl4_metadata
  lvl4_subs <- subset(
    sc_obj,
    subset = annotation_deconv %in% unique(lvl4_metadata$annotation_level_3))
  
  #### Extract marker genes from each cluster ####
  Seurat::Idents(object = lvl4_subs) <- lvl4_subs@meta.data[, "annotation_deconv"]
  
  markers_lvl4 <- Seurat::FindAllMarkers(
    object = lvl4_subs, 
    assay = "RNA",
    slot = "data",
    verbose = TRUE, 
    only.pos = TRUE,
    max.cells.per.ident = 200) 
  
  saveRDS(object = markers_lvl4, file = markers_lvl4_out)
}

DT::datatable(markers_lvl4, filter = "top")
```

Join markers from both runs. Since we still want to keep the epithelial cells markers in the lower level markers we will add them to each one.
```{r}
# Keep markers for major cell types
# gene_vec <- c("BCL6", "PRDM1", "FDCSP")
markers_lvl1_sub <- markers_lvl1 %>% dplyr::filter(cluster != "epithelial") %>%
    dplyr::filter( ( avg_log2FC > 1 | pct.1 - pct.2 > 0.75) &
                   p_val_adj < 0.01 & pct.1 > 0.3)
                   # | gene %in% gene_vec)

# Keep markers for general epithelial T cells
markers_lvl1_epith <- markers_lvl1 %>%
  dplyr::filter(cluster == "epithelial") %>%
  # Filter to get CD4 specific markers
  # We are keeping those genes with a logFC > 1 or a pct.1 - pct.2 > 0.75 which
  # have a p_val_adj < 0.01 & pct.1 > 0.6 while keeping CD4
  dplyr::filter( ( avg_log2FC > 1 | pct.1 - pct.2 > 0.75) &
                   p_val_adj < 0.01 & pct.1 > 0.6)
```

Add general Epithelial cell markers to the specific cell types
```{r}

markers_lvl4_update <- lapply(as.character(unique(markers_lvl4$cluster)), function (i) {
  # Filter level 5 clusters
  tmp <- markers_lvl4 %>%
    # dplyr::filter(avg_log2FC > 0.75 | pct.1 - pct.2 > 0.4) %>%
    # dplyr::filter(avg_log2FC > 0.5 | pct.1 - pct.2 > 0.4) %>%
    dplyr::filter(cluster == i) %>%
    # Remove ribosomal genes
    dplyr::filter(!stringr::str_detect(string = gene,
                                      pattern = "^RPS|^RPL"))
    
  # Add general CD4 markers
  # tmp %>%
  #   dplyr::bind_rows(markers_lvl1_epith) %>%
  #   dplyr::mutate(cluster = i)
}) %>%
  dplyr::bind_rows()

DT::datatable(markers_lvl4_update, filter = "top")
```

Combine level-1 and level-4 markers
```{r}
all_markers <- dplyr::bind_rows(markers_lvl1_sub, markers_lvl4_update) %>%
  dplyr::arrange(dplyr::desc(avg_log2FC)) %>%
  dplyr::group_by(cluster, gene) %>%
  dplyr::filter(dplyr::row_number() == 1)

DT::datatable(all_markers, filter = "top")
```


Subset SC object so all the cells from the same cell type to come from the same batch
```{r}
# subset most representative sample(s)
ns <- with(sc_obj@meta.data, table(gem_id, annotation_deconv))
ns <- as.matrix(unclass(ns))
id <- apply(ns, 2, function(.) {
    n <- 0
    s <- c()
    while (n < 20 & length(.) > 0) {
        i <- which.max(.)
        s <- c(s, names(i))
        n <- n + .[i]
        . <- .[-i]
    }
    return(s)
})

# downsample to at most 50 cells per population
cs <- mapply(function(s, k) {
    cs <- with(sc_obj@meta.data, which(gem_id %in% s & annotation_deconv == k))
    sample(cs, min(20, length(cs)))
}, s = id, k = names(id))
sc_sub <- sc_obj[, unlist(cs)]

table(sc_sub@meta.data$annotation_deconv)
```

### SPOTlight deconvolution
2nd run deconvolution
```{r}
Seurat::Idents(object = sc_sub) <- as.character(sc_sub@meta.data$annotation_deconv)

spotlight_ls <- spotlight_deconvolution(
  se_sc = sc_sub,
  counts_spatial = sp_obj@assays$Spatial@counts,
  clust_vr = "annotation_deconv",
  cluster_markers = all_markers,
  cl_n = 100,
  hvg = 0,
  ntop = NULL,
  transf = "uv",
  method = "nsNMF",
  min_cont = 0)
```

## Save data
```{r}
decon_fn <- "{epithelium}/{robj_dir}/spotlight_ls_epithelium.rds" %>%
  glue::glue() %>% 
  here::here()
# if (! file.exists(decon_fn) ) {
saveRDS(
  object = spotlight_ls,
  file = decon_fn)
# } else {
#   spotlight_ls <- readRDS(file = here::here(glue::glue("{epithelium}/{robj_dir}/spotlight_ls_cd4.rds")))
# }
```

3rd Check Topic profiles
```{r fig.height=8, fig.width=8}
nmf_mod_ls <- spotlight_ls[[1]]
nmf_mod <- nmf_mod_ls[[1]]

h <- NMF::coef(nmf_mod)
rownames(h) <- paste("Topic", 1:nrow(h), sep = "_")
topic_profile_plts <- SPOTlight::dot_plot_profiles_fun(
  h = h,
  train_cell_clust = nmf_mod_ls[[2]])

topic_profile_plts[[2]] +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90), 
                 axis.text = ggplot2::element_text(size = 12))
```

## Visualize deconvolution results
```{r}
decon_mtrx <- spotlight_ls[[2]]
decon_mtrx <- decon_mtrx[, colnames(decon_mtrx) != "res_ss"]

# Set as 0 cell types predicted to be under 1 % of the spot
# decon_mtrx[decon_mtrx < 0.03] <- 0
```

Change column names
```{r}
nm_df <- data.frame(
  mod_nm = stringr::str_replace_all(string = unique(all_markers$cluster),
                                    pattern = "[[:punct:]]|\\s|\\+" ,
                                    replacement = "."),
  plt_nm = unique(all_markers$cluster)
  )

# Save nm_df
"{epithelium}/{robj_dir}/epithelium_nm_df.rds" %>%
  glue::glue() %>% 
  here::here() %>%
  saveRDS(
    object = nm_df,
    file = .)


# new_cn <- data.frame(mod_nm = colnames(decon_mtrx)) %>%
#   dplyr::left_join(nm_df, by = "mod_nm") %>%
#   # Central.Mem.PASK. fives some trouble because it only changes between + an -
#   # negative goes first and distinct solves it automatically
#   dplyr::distinct() %>%
#   dplyr::pull(plt_nm)
# 
# colnames(decon_mtrx) <- new_cn
```

We are going to add the deconvolution to the Seurat object.
```{r}
rownames(decon_mtrx) <- rownames(sp_obj@meta.data)

sp_obj@meta.data <- sp_obj@meta.data %>%
  tibble::rownames_to_column("barcode") %>%
  dplyr::left_join(data.frame(decon_mtrx) %>%
                     tibble::rownames_to_column("barcode"), by = "barcode") %>%
  tibble::column_to_rownames("barcode")
```

Look at cells topic profile
```{r}
basis_spotlight <- data.frame(NMF::basis(spotlight_ls[[1]][[1]]))

train_labs <- spotlight_ls[[1]][[2]]
colnames(basis_spotlight) <- unique(stringr::str_wrap(train_labs, width = 30))

basis_spotlight[basis_spotlight < 0.0000001] <- 0

DT::datatable(basis_spotlight, filter = "top")
```

#### Cell type location
Look at the location of each cell type in each slice separately
```{r fig.height=25, fig.width=25}
# Iterate over cell types
ct <- colnames(decon_mtrx)
# Seurat::Idents(sp_obj) <- sp_obj@meta.data$gem_id
sp_sub <- sp_obj[, sp_obj@meta.data$gem_id == sample_id]
sp_sub@images <- sp_sub@images[Seurat::Images(sp_sub) == sample_id]

ct_plt_ls <- lapply(ct, function(i) {
  tmp_plt <- Seurat::SpatialFeaturePlot(object = sp_sub,
                                        features = i,
                                        alpha = c(0, 1)) +
    ggplot2::scale_fill_gradientn(
      colors = heat.colors(10, rev = TRUE)) +
    ggplot2::scale_alpha(range = c(0, 1)) +
    ggplot2::labs(title = stringr::str_wrap(string = i,
                                   width = 25),
         fill = "") +
    ggplot2::theme(
      plot.title = ggplot2::element_text(
        hjust = 0.5,
        size = 20,
        face = "bold"))
  
  if (sum(sp_sub@meta.data[, i]) == 0) {
    tmp_plt <- suppressMessages(tmp_plt + ggplot2::scale_alpha(range = c(0,0)))
  }
  
  return(tmp_plt)
})

(plt_arr <- cowplot::plot_grid(
  plotlist = ct_plt_ls,
  axis = "trbl",
  align = "hv",
  nrow = 5,
  ncol = 5))

"{epithelium}/{plt_dir}/cell_type_location_epithelium_{donor_id}.pdf" %>%
  glue::glue() %>%
  here::here() %>%
  cowplot::save_plot(
    filename = .,
    plot = plt_arr,
    base_height = 25,
    base_width = 25)
```

## Session Info
```{r}
sessionInfo()
```
