---
author: "Marc Elosua-Bayes"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
params: 
    sample_id: "Default!"
    donor_id: "Default!"
title: "Myeloid MAGIC visualization"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# knitr::opts_knit$set(root.dir = "..")
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message = FALSE, warning = FALSE)
options(width = 1200)
```

## Introduction
In this Rmarkdown document we are going to load the MAGIC-denoised data to better visualize genes and ease with the annotation when using specific marker genes.
[MAGIC](https://magic.readthedocs.io/en/stable/) was developed by Smita Krishnaswamy's lab to try to fill in the drop out reads in the spots. `MAGIC` is a Markov Affinity-based Graph Imputation of Cells used for denoising high-dimensional data most commonly applied to single-cell RNA sequencing data. MAGIC learns the manifold data, using the resultant graph to smooth the features and restore the structure of the data based on their k-nearest neighbors. 

## Libraries
```{r}
library(Seurat)
library(dplyr)
library(ggplot2)
library(SPATA2)
library(UCell)
library(stringr)
```

## Parameters
```{r}
set.seed(123)
source(here::here("misc/paths.R"))
source(here::here("utils/bin.R"))

"{myeloid}/{plt_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = .,
             showWarnings = FALSE,
             recursive = TRUE)

"{myeloid}/{robj_dir}" %>%
  glue::glue() %>%
  here::here() %>%
  dir.create(path = .,
             showWarnings = FALSE,
             recursive = TRUE)
```

## Load data
The data used in this Rmarkdown document comes from **03-clustering_integration.Rmd** where the data was integrated.
```{r}
merged_se <- "{anot}/{robj_dir}/integrated_spatial_annot.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)
```

Load MAGIC data from the script *MAGIC_denoising-myeloid.Rmd*
```{r}
magic_df <- "{myeloid}/{robj_dir}/MAGIC-mtrx.rds" %>%
  glue::glue() %>%
  here::here() %>%
  readRDS(file = .)

# create a new assay to store ADT information
magic_assay <- CreateAssayObject(counts = as.matrix(magic_df))

# Subset merged_se to those barcodes used
merged_se <- merged_se[, colnames(magic_df)]

# add this assay to the previously created Seurat object
merged_se[["MAGIC_Spatial"]] <- magic_assay
```

## Analysis

[A Spatially Resolved Dark- versus Light-Zone Microenvironment Signature Subdivides Germinal Center-Related Aggressive B Cell Lymphomas](10.1016/j.isci.2020.101562)
- **Dark Zone**: OAZ1, AICDA, H3, MKI67, POLH
- **Light Zone**: LAG3, ITGB8, PDCD1, TIGIT, BCL2, PECAM1, LY6E, B7-H3 (CD276), HLA-DRB1, PSMB10, TNF, ARG1, HLA-E, STAT1

```{r}
"{myeloid}/gene_dict_myeloid.R" %>%
  glue::glue() %>%
  here::here() %>%
  source(file = .)
gene_vec <- rownames(magic_df)
```

### Marker gene location
Look at the location where the genes of interest are expressed
```{r fig.width=20, fig.height=30}
Seurat::DefaultAssay(merged_se) <- "MAGIC_Spatial"

# Iterate over each image
lapply(id_sp_df$gem_id, function(i) {
  print(i)

  gene_plt <- Seurat::SpatialFeaturePlot(
    object = merged_se,
    features = myeloid_vec,
    alpha = c(0, 1),
    ncol = 7,
    images = i)

  # Save plot
  "{myeloid}/{plt_dir}/magic_myeloid_markers_{i}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = gene_plt,
      base_height = 45,
      base_width = 40)
})
```

Now with the log-norm expression
```{r fig.width=20, fig.height=30}
Seurat::DefaultAssay(merged_se) <- "Spatial"

lapply(id_sp_df$gem_id, function(i) {
  # Iterate over each image
  gene_plt <- Seurat::SpatialFeaturePlot(
    object = merged_se,
    features = myeloid_vec,
    alpha = c(0, 1),
    ncol = 7,
    images = i)

  "{myeloid}/{plt_dir}/lognorm_myeloid_markers_{i}.pdf" %>%
    glue::glue() %>%
    here::here() %>%
    cowplot::save_plot(
      filename = .,
      plot = gene_plt,
      base_height = 45,
      base_width = 40)
})
```

```{r}
se_sub <- subset(merged_se, subset = gem_id == "esvq52_nluss5")
se_sub
se_sub@images <- se_sub@images[Seurat::Images(se_sub) == "esvq52_nluss5"]
```

### Correlation matrix
Since we are working with sample esvq52_nluss5 in this example we will limit the correlation plot to this slide.
```{r fig.height=15, fig.width=15}
(cor_mtrx <- SCrafty::correlation_heatmap( 
  se = se_sub,
  genes = gene_vec,
  assay = "MAGIC_Spatial",
  slot = "data"))

"{myeloid}/{plt_dir}/magic_cor-mtrx_markers.pdf" %>%
  glue::glue() %>%
  here::here() %>%
  cowplot::save_plot(
    filename = .,
    plot = cor_mtrx,
    base_height = 15,
    base_width = 15)

  # Correlation with lognorm expression
cor_log <- SCrafty::correlation_heatmap( 
  se = se_sub,
  genes = gene_vec,
  assay = "Spatial",
  slot = "data")

"{myeloid}/{plt_dir}/lognorm_cor-mtrx_markers.pdf" %>%
  glue::glue() %>%
  here::here() %>%
  cowplot::save_plot(
    filename = .,
    plot = cor_log,
    base_height = 9,
    base_width = 10)
```

Look at them side by side
```{r fig.width=30, fig.height=15}
cor_mtrx + cor_log
```

## Session Info
```{r}
sessionInfo()
```

